{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import glob\n",
    "from scipy import stats\n",
    "from sklearn.metrics.pairwise import cosine_similarity\n",
    "import networkx as nx\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Helper function to find files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def abs_path(target_name, directory_level): \n",
    "    \"\"\"\n",
    "Returns absolute file path of target name in working directory.\n",
    "\n",
    "Arguments:\n",
    "    target_name (str): Name of file or folder to find.\n",
    "    directory_level (str): Level of os search, either File or Folder.   \n",
    "    \"\"\"\n",
    "    #Find the relative working directory of the script\n",
    "    wk_dir = os.path.dirname(os.path.realpath('__file__'))\n",
    "    \n",
    "    if directory_level == \"File\":\n",
    "        #Absolute file path\n",
    "        for root, dirs, files in os.walk(wk_dir):\n",
    "            for name in files:\n",
    "                if target_name == name:\n",
    "                    target_path = (os.path.abspath(os.path.join(root, name))) \n",
    "             \n",
    "    #Absolute file path\n",
    "    if directory_level == \"Directory\":\n",
    "        for root, dirs, files in os.walk(wk_dir):\n",
    "            for name in dirs:\n",
    "                if target_name == name:\n",
    "                    target_path = (os.path.abspath(os.path.join(root, name))) \n",
    "    \n",
    "    return target_path"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mutation burden of simulated sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mut_burden(cancer_type, simulation_type, gen_start, gen_end):\n",
    "\n",
    "    mutation_burden = pd.DataFrame(index = range(gen_start, gen_end+1), columns = ['sbs', 'dbs', 'ins', 'del'])\n",
    "\n",
    "    for gen in range(gen_start, gen_end+1):\n",
    "        try:\n",
    "            sbs_burden = pd.read_csv(abs_path(cancer_type + '_' + str(simulation_type) + '_Stage_Lineage_' + str(gen) + '_' + \"sbs\" + '_freq_table.csv', 'File'))['Frequency'].sum()\n",
    "            dbs_burden = pd.read_csv(abs_path(cancer_type + '_' + str(simulation_type) + '_Stage_Lineage_' + str(gen) + '_' + \"dbs\" + '_freq_table.csv', 'File'))['Frequency'].sum()\n",
    "            ins_burden = pd.read_csv(abs_path(cancer_type + '_' + str(simulation_type) + '_Stage_Lineage_' + str(gen) + '_' + \"ins\" + '_freq_table.csv', 'File'))['Frequency'].sum()\n",
    "            del_burden = pd.read_csv(abs_path(cancer_type + '_' + str(simulation_type) + '_Stage_Lineage_' + str(gen) + '_' + \"del\" + '_freq_table.csv', 'File'))['Frequency'].sum()\n",
    "\n",
    "            mutation_burden.iloc[list(range(gen_start, gen_end+1)).index(gen), 0] = sbs_burden\n",
    "            mutation_burden.iloc[list(range(gen_start, gen_end+1)).index(gen), 1] = dbs_burden\n",
    "            mutation_burden.iloc[list(range(gen_start, gen_end+1)).index(gen), 2] = ins_burden\n",
    "            mutation_burden.iloc[list(range(gen_start, gen_end+1)).index(gen), 3] = del_burden\n",
    "        except:\n",
    "            print(gen)\n",
    "        \n",
    "\n",
    "    return mutation_burden"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "1\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "10\n",
      "11\n",
      "12\n",
      "13\n",
      "14\n",
      "15\n",
      "16\n",
      "17\n",
      "18\n",
      "19\n",
      "20\n",
      "21\n",
      "22\n",
      "23\n",
      "24\n",
      "25\n",
      "26\n",
      "27\n",
      "28\n",
      "29\n",
      "30\n",
      "31\n",
      "32\n",
      "33\n",
      "34\n",
      "35\n",
      "36\n",
      "37\n",
      "38\n",
      "39\n",
      "40\n",
      "41\n",
      "42\n",
      "43\n",
      "44\n",
      "45\n",
      "46\n",
      "47\n",
      "48\n",
      "49\n",
      "50\n",
      "51\n",
      "52\n",
      "53\n",
      "54\n",
      "55\n",
      "56\n",
      "57\n",
      "58\n",
      "59\n",
      "60\n",
      "61\n",
      "62\n",
      "63\n",
      "64\n",
      "65\n",
      "66\n",
      "67\n",
      "68\n",
      "69\n",
      "70\n",
      "71\n",
      "72\n",
      "73\n",
      "74\n",
      "75\n",
      "76\n",
      "77\n",
      "78\n",
      "79\n",
      "80\n",
      "81\n",
      "82\n",
      "83\n",
      "84\n",
      "85\n",
      "86\n",
      "87\n",
      "88\n",
      "89\n",
      "90\n",
      "91\n",
      "92\n",
      "93\n",
      "94\n",
      "95\n",
      "96\n",
      "97\n",
      "98\n",
      "99\n",
      "100\n",
      "101\n",
      "102\n",
      "103\n",
      "104\n",
      "105\n",
      "106\n",
      "107\n",
      "108\n",
      "109\n",
      "110\n",
      "111\n",
      "112\n",
      "113\n",
      "114\n",
      "115\n",
      "116\n",
      "117\n",
      "118\n",
      "119\n",
      "120\n",
      "121\n",
      "122\n",
      "123\n",
      "124\n",
      "125\n",
      "126\n",
      "127\n",
      "128\n",
      "129\n",
      "130\n",
      "131\n",
      "132\n",
      "133\n",
      "134\n",
      "135\n",
      "136\n",
      "137\n",
      "138\n",
      "139\n",
      "140\n",
      "141\n",
      "142\n",
      "143\n",
      "144\n",
      "145\n",
      "146\n",
      "147\n",
      "148\n",
      "149\n",
      "150\n",
      "151\n",
      "152\n",
      "153\n",
      "154\n",
      "155\n",
      "156\n",
      "157\n",
      "158\n",
      "159\n",
      "160\n",
      "161\n",
      "162\n",
      "163\n",
      "164\n",
      "165\n",
      "166\n",
      "167\n",
      "168\n",
      "169\n",
      "170\n",
      "171\n",
      "172\n",
      "173\n",
      "174\n",
      "175\n",
      "176\n",
      "177\n",
      "178\n",
      "179\n",
      "180\n",
      "181\n",
      "182\n",
      "183\n",
      "184\n",
      "185\n",
      "186\n",
      "187\n",
      "188\n",
      "189\n",
      "190\n",
      "191\n",
      "192\n",
      "193\n",
      "194\n",
      "195\n",
      "196\n",
      "197\n",
      "198\n",
      "199\n",
      "200\n",
      "201\n",
      "202\n",
      "203\n",
      "204\n",
      "205\n",
      "206\n",
      "207\n",
      "208\n",
      "209\n",
      "210\n",
      "211\n",
      "212\n",
      "213\n",
      "214\n",
      "215\n",
      "216\n",
      "217\n",
      "218\n",
      "219\n",
      "220\n",
      "221\n",
      "222\n",
      "223\n",
      "224\n",
      "225\n",
      "226\n",
      "227\n",
      "228\n",
      "229\n",
      "230\n",
      "231\n",
      "232\n",
      "233\n",
      "234\n",
      "235\n",
      "236\n",
      "237\n",
      "238\n",
      "239\n",
      "240\n",
      "241\n",
      "242\n",
      "243\n",
      "244\n",
      "245\n",
      "246\n",
      "247\n",
      "248\n",
      "249\n",
      "250\n",
      "251\n",
      "252\n",
      "253\n",
      "254\n",
      "255\n",
      "256\n",
      "257\n",
      "258\n",
      "259\n",
      "260\n",
      "261\n",
      "262\n",
      "263\n",
      "264\n",
      "265\n",
      "266\n",
      "267\n",
      "268\n",
      "269\n",
      "270\n",
      "271\n",
      "272\n",
      "273\n",
      "274\n",
      "275\n",
      "276\n",
      "277\n",
      "278\n",
      "279\n",
      "280\n",
      "281\n",
      "282\n",
      "283\n",
      "284\n",
      "285\n",
      "286\n",
      "287\n",
      "288\n",
      "289\n",
      "290\n",
      "291\n",
      "292\n",
      "293\n",
      "294\n",
      "295\n",
      "296\n",
      "297\n",
      "298\n",
      "299\n",
      "300\n",
      "301\n",
      "302\n",
      "303\n",
      "304\n",
      "305\n",
      "306\n",
      "307\n",
      "308\n",
      "309\n",
      "310\n",
      "311\n",
      "312\n",
      "313\n",
      "314\n",
      "315\n",
      "316\n",
      "317\n",
      "318\n",
      "319\n",
      "320\n",
      "321\n",
      "322\n",
      "323\n",
      "324\n",
      "325\n",
      "326\n",
      "327\n",
      "328\n",
      "329\n",
      "330\n",
      "331\n",
      "332\n",
      "333\n",
      "334\n",
      "335\n",
      "336\n",
      "337\n",
      "338\n",
      "339\n",
      "340\n",
      "341\n",
      "342\n",
      "343\n",
      "344\n",
      "345\n",
      "346\n",
      "347\n",
      "348\n",
      "349\n",
      "350\n",
      "351\n",
      "352\n",
      "353\n",
      "354\n",
      "355\n",
      "356\n",
      "357\n",
      "358\n",
      "359\n",
      "360\n",
      "361\n",
      "362\n",
      "363\n",
      "364\n",
      "365\n",
      "366\n",
      "367\n",
      "368\n",
      "369\n",
      "370\n",
      "371\n",
      "372\n",
      "373\n",
      "374\n",
      "375\n",
      "376\n",
      "377\n",
      "378\n",
      "379\n",
      "380\n",
      "381\n",
      "382\n",
      "383\n",
      "384\n",
      "385\n",
      "386\n",
      "387\n",
      "388\n",
      "389\n",
      "390\n",
      "391\n",
      "392\n",
      "393\n",
      "394\n",
      "395\n",
      "396\n",
      "397\n",
      "398\n",
      "399\n",
      "400\n",
      "401\n",
      "402\n",
      "403\n",
      "404\n",
      "405\n",
      "406\n",
      "407\n",
      "408\n",
      "409\n",
      "410\n",
      "411\n",
      "412\n",
      "413\n",
      "414\n",
      "415\n",
      "416\n",
      "417\n",
      "418\n",
      "419\n",
      "420\n",
      "421\n",
      "422\n",
      "423\n",
      "424\n",
      "425\n",
      "426\n",
      "427\n",
      "428\n",
      "429\n",
      "430\n",
      "431\n",
      "432\n",
      "433\n",
      "434\n",
      "435\n",
      "436\n",
      "437\n",
      "438\n",
      "439\n",
      "440\n",
      "441\n",
      "442\n",
      "443\n",
      "444\n",
      "445\n",
      "446\n",
      "447\n",
      "448\n",
      "449\n",
      "450\n",
      "451\n",
      "452\n",
      "453\n",
      "454\n",
      "455\n",
      "456\n",
      "457\n",
      "458\n",
      "459\n",
      "460\n",
      "461\n",
      "462\n",
      "463\n",
      "464\n",
      "465\n",
      "466\n",
      "467\n",
      "468\n",
      "469\n",
      "470\n",
      "471\n",
      "472\n",
      "473\n",
      "474\n",
      "475\n",
      "476\n",
      "477\n",
      "478\n",
      "479\n",
      "480\n",
      "481\n",
      "482\n",
      "483\n",
      "484\n",
      "485\n",
      "486\n",
      "487\n",
      "488\n",
      "489\n",
      "490\n",
      "491\n",
      "492\n",
      "493\n",
      "494\n",
      "495\n",
      "496\n",
      "497\n",
      "498\n",
      "499\n"
     ]
    }
   ],
   "source": [
    "foo = mut_burden(\"Liver-HCC\", \"End\", 0,499)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sbs</th>\n",
       "      <th>dbs</th>\n",
       "      <th>ins</th>\n",
       "      <th>del</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sbs  dbs  ins  del\n",
       "0  NaN  NaN  NaN  NaN\n",
       "1  NaN  NaN  NaN  NaN\n",
       "2  NaN  NaN  NaN  NaN\n",
       "3  NaN  NaN  NaN  NaN\n",
       "4  NaN  NaN  NaN  NaN\n",
       "5  NaN  NaN  NaN  NaN\n",
       "6  NaN  NaN  NaN  NaN\n",
       "7  NaN  NaN  NaN  NaN\n",
       "8  NaN  NaN  NaN  NaN\n",
       "9  NaN  NaN  NaN  NaN"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "foo.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Reference data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Input sequence file path\n",
    "\"\"\"\n",
    "input_file_path = abs_path(\"Homo_sapiens.GRCh38.dna.chromosome.22.fasta\", \"File\")\n",
    "# print(input_sequence)\n",
    "#input_sequence = seq_slice(input_file_path, 26100000, 27100000)\n",
    "sig_weights =  pd.read_csv(abs_path(\"timelines_sigWeights.txt\", \"File\"), sep='\\t')\n",
    "\"\"\"\n",
    "Read in input reference files:\n",
    "data = Pandas Dataframe of number of mutations attributed to each signature for each sample.\n",
    "cancer_type = List of unique cancer types.\n",
    "prop_data = Pandas Dataframe of proportion of 96 type mutation classification for each signature.\n",
    "\"\"\"\n",
    "\n",
    "\"\"\"\n",
    "SBS input reference files\n",
    "\"\"\"\n",
    "sbs_num_file_path = abs_path(\"PCAWG_sigProfiler_SBS_signatures_in_samples.csv\",\"File\")\n",
    "sbs_num_data = pd.read_csv(sbs_num_file_path)\n",
    "sbs_num_data.sort_values(by='Cancer Types', axis=0, inplace=True)\n",
    "sbs_num_data.set_index(keys=['Cancer Types'], drop=False,inplace=True)\n",
    "\n",
    "sbs_prop_file_path = abs_path(\"sigProfiler_SBS_signatures.csv\",\"File\")\n",
    "sbs_prop_data = pd.read_csv(sbs_prop_file_path)\n",
    "  \n",
    "\n",
    "\"\"\"\n",
    "DBS input reference files\n",
    "\"\"\"\n",
    "dbs_num_file_path = abs_path(\"PCAWG_sigProfiler_DBS_signatures_in_samples.csv\",\"File\")\n",
    "dbs_num_data = pd.read_csv(dbs_num_file_path)\n",
    "dbs_num_data.sort_values(by='Cancer Types', axis=0, inplace=True)\n",
    "dbs_num_data.set_index(keys=['Cancer Types'], drop=False,inplace=True)\n",
    "\n",
    "dbs_prop_file_path = abs_path(\"sigProfiler_DBS_signatures.csv\",\"File\")\n",
    "dbs_prop_data = pd.read_csv(dbs_prop_file_path)\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "Insertion/Deletion input reference files\n",
    "\"\"\"\n",
    "id_num_file_path = abs_path(\"PCAWG_SigProfiler_ID_signatures_in_samples.csv\",\"File\")\n",
    "id_num_data = pd.read_csv(id_num_file_path)\n",
    "id_num_data.sort_values(by='Cancer Types', axis=0, inplace=True)\n",
    "id_num_data.set_index(keys=['Cancer Types'], drop=False,inplace=True)\n",
    "\n",
    "id_prop_file_path = abs_path(\"sigProfiler_ID_signatures.csv\",\"File\")\n",
    "id_prop_data = pd.read_csv(id_prop_file_path)\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "Expected Frequency Absolute Folder Paths for SBS, DBS and ID\n",
    "\"\"\"\n",
    "sbs_freq_folder_path = abs_path(\"SBS_Expected_Frequency\", \"Directory\")\n",
    "dbs_freq_folder_path = abs_path(\"DBS_Expected_Frequency\", \"Directory\")\n",
    "id_freq_folder_path = abs_path(\"ID_Expected_Frequency\", \"Directory\")\n",
    "\n",
    "#%%\n",
    "\n",
    "\n",
    "def sig_proportion(cancer_type, num_data, std_outlier):\n",
    "    \"\"\"\n",
    "Find the proportions of different combinations of signatures among all samples\n",
    "Output proportion of samples with each combination of signatures. \n",
    "\n",
    "Arguments:\n",
    "    cancer_type (str): Cancer type \n",
    "    num_data (dataframe): Signature data\n",
    "    std_outlier (float): Parameter to exclude outliers by degrees of standard deviation \n",
    "    \"\"\"\n",
    "    #Setup dataframe with signature data\n",
    "    n_mutation_df = num_data.loc[cancer_type,\"Sample Names\":]\n",
    "    n_mutation_df.set_index(\"Sample Names\", inplace=True)\n",
    "    del n_mutation_df['Accuracy']\n",
    "    \n",
    "    cols = n_mutation_df.columns\n",
    "    \n",
    "    #Find contributing signature combinations for each sequence\n",
    "    sig_combos = list(n_mutation_df.apply(lambda x: x > 0).apply(lambda x: list(n_mutation_df.columns[x.values]), axis=1))\n",
    "  \n",
    "    #Find the set of unique contributing signature combinations \n",
    "    unique_sig_combos = [list(x) for x in set(tuple(x) for x in sig_combos)]\n",
    "    \n",
    "    \n",
    "    #All outliers found for each sequence in no particular order\n",
    "    net_outlier_sample = []\n",
    "    \n",
    "    for index in unique_sig_combos:\n",
    "        combo_df = n_mutation_df[index][(n_mutation_df[index] != 0).all(1)]\n",
    "        filter_combo_df = combo_df[~(np.abs(combo_df-combo_df.mean()) > (std_outlier*combo_df.std()))]\n",
    "\n",
    "        for sig in index:\n",
    "            net_outlier_sample.append(filter_combo_df[filter_combo_df[sig].isnull()].index.tolist())\n",
    "    \n",
    "    #Set of unique outlier sequences\n",
    "    set_outlier = list(set(x for l in net_outlier_sample for x in l))\n",
    "    \n",
    "    #print(\"Removed the following outliers from the signature combination probability matrix:\" + str(set_outlier))\n",
    "\n",
    "    #Drop outlier sequences \n",
    "    n_mutation_df.drop(set_outlier,inplace=True)\n",
    "    \n",
    "    #Dataframe of the proportion of a set of signatures contribution out of all possible combinations\n",
    "    bt=n_mutation_df.apply(lambda x: x > 0)\n",
    "    \n",
    "    sig_prop = ( bt.apply( lambda x: (cols[x.values]) , axis=1).value_counts() ) / len(bt)\n",
    "    \n",
    "    return sig_prop\n",
    "\n",
    "\n",
    "\n",
    "def outlier_detection(cancer_type, num_data, std_outlier):\n",
    "    \"\"\"\n",
    "Identifies outliers in the dataframe, returns list of outliers\n",
    "\n",
    "Arguments;\n",
    "    cancer_type (str): Cancer type\n",
    "    num_data (dataframe): Signature data\n",
    "    std_outlier: Parameter to exclude outliers by degrees of standard deviation \n",
    "    \"\"\" \n",
    "    #Setup dataframe with signature data\n",
    "    n_mutation_df = num_data.loc[cancer_type,\"Sample Names\":]\n",
    "    n_mutation_df.set_index(\"Sample Names\", inplace=True)\n",
    "    del n_mutation_df['Accuracy']\n",
    "    \n",
    "    #Find contributing signature combinations for each sequence\n",
    "    sig_combos = list(n_mutation_df.apply(lambda x: x > 0).apply(lambda x: list(n_mutation_df.columns[x.values]), axis=1))\n",
    "  \n",
    "    #Find the set of unique contributing signature combinations \n",
    "    unique_sig_combos = [list(x) for x in set(tuple(x) for x in sig_combos)]\n",
    "    \n",
    "    #All outliers found for each sequence in no particular order\n",
    "    net_outlier_sample = []\n",
    "    \n",
    "    for index in unique_sig_combos:\n",
    "        combo_df = n_mutation_df[index][(n_mutation_df[index] != 0).all(1)]\n",
    "        filter_combo_df = combo_df[~(np.abs(combo_df-combo_df.mean()) > (std_outlier*combo_df.std()))]\n",
    "\n",
    "        for sig in index:\n",
    "            net_outlier_sample.append(filter_combo_df[filter_combo_df[sig].isnull()].index.tolist())\n",
    "    \n",
    "    #Set of unique outlier sequences\n",
    "    set_outlier = list(set(x for l in net_outlier_sample for x in l))\n",
    "    \n",
    "    return set_outlier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ref_data(input_name, num=False, prop=False):\n",
    "    \"\"\"\n",
    "Reads in reference datasets from PCAWG studies.\n",
    "\n",
    "Arguments:\n",
    "    input_name (str): Name of csv file \n",
    "    num (bool): Set true to return dataframe of PCAWG SigProfiler Signatures in Samples \n",
    "    prop (bool): Set true to return dataframe of SigProfiler Signature Contributions\n",
    "    \n",
    "Note: Must set only one of num or prop arguments to True. \n",
    "    \"\"\"\n",
    "    if prop == num:\n",
    "        raise ValueError(\"Please set only one argument between the arguments num and prop to True.\")\n",
    "        \n",
    "    if num is True:\n",
    "        \n",
    "        num_file_path = abs_path(input_name, \"File\")\n",
    "        num_data = pd.read_csv(num_file_path)\n",
    "        num_data.sort_values(by='Cancer Types', axis=0, inplace=True)\n",
    "        num_data.set_index(keys=['Cancer Types'], drop=False,inplace=True)\n",
    "        \n",
    "        return num_data\n",
    "\n",
    "    if prop is True:\n",
    "            \n",
    "        prop_file_path = abs_path(input_name, input_type )\n",
    "        prop_data = pd.read_csv(prop_file_path)\n",
    "        \n",
    "        return prop_data\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mut_mean(cancer_type, num_data, freq_folder_path, mut_type, std_outlier):\n",
    "    \"\"\"\n",
    "Output average number of 96 type mutations among all samples for specified cancer type.\n",
    "\n",
    "Arguments:\n",
    "    cancer_type (str): Cancer type\n",
    "    num_data (dataframe): Signature data\n",
    "    freq_folder_path (str): Folder with expected mutation frequencies for each sample\n",
    "    mut_type (str): \"SBS\", \"DBS\", \"ID\"\n",
    "    std_outlier (float): Parameter to exclude outliers by degrees of standard deviation \n",
    "    \"\"\"\n",
    "    if mut_type == \"SBS\":\n",
    "         \n",
    "        outlier_list = outlier_detection(cancer_type, num_data, std_outlier)\n",
    "        outlier_paths = [abs_path(\"SBS_\" + cancer_type + \"_\" + outlier + \".csv\", \"File\") for outlier in outlier_list]\n",
    "        file_paths = glob.glob(str(sbs_freq_folder_path) + \"/SBS_\" + cancer_type + '_*' + '.csv')\n",
    "        filtered_file_paths = [x for x in file_paths if x not in outlier_paths]\n",
    "        \n",
    "\n",
    "        #print(\"Removed \" + str(len(file_paths)-len(filtered_file_paths)) + \" outliers from the average mutation frequency matrix\")\n",
    "        #print(outlier_list)\n",
    "        \n",
    "        og_df = pd.read_csv(filtered_file_paths[0])\n",
    "        mean_df = og_df.loc[:, 'SBS1':]\n",
    "    \n",
    "        for file_index in range(1, len(file_paths)):\n",
    "            raw_mean = pd.read_csv(file_paths[file_index])\n",
    "            slice_read_df = raw_mean.loc[:, 'SBS1':]\n",
    "            mean_df += slice_read_df\n",
    "            \n",
    "        df_averages = mean_df/len(file_paths) \n",
    "        mut_subtypes = og_df.loc[:, 'Type': 'SubType']\n",
    "        average_freq_df = pd.concat([mut_subtypes, df_averages], axis=1)\n",
    "        \n",
    "        return average_freq_df\n",
    "  \n",
    "    if mut_type == \"DBS\":\n",
    "        \n",
    "        outlier_list = outlier_detection(cancer_type, num_data, std_outlier)\n",
    "        outlier_paths = [abs_path(\"DBS_\" + cancer_type + \"_\" + outlier + \".csv\", \"File\") for outlier in outlier_list]\n",
    "        file_paths = glob.glob(str(dbs_freq_folder_path)+ \"/DBS_\" + cancer_type + '_*' + '.csv')\n",
    "        filtered_file_paths = [x for x in file_paths if x not in outlier_paths]\n",
    "        \n",
    "        #print(\"Dropped \" + str(len(file_paths)-len(filtered_file_paths)) + \" outliers!\")\n",
    "        #print(\"\")\n",
    "        #print(outlier_list)\n",
    "        \n",
    "        og_df = pd.read_csv(filtered_file_paths[0], index_col=0)\n",
    "        mean_df = og_df.loc[:, 'DBS1':]\n",
    "        \n",
    "        for file_index in range(1, len(file_paths)):\n",
    "            raw_mean = pd.read_csv(file_paths[file_index])\n",
    "            slice_read_df = raw_mean.loc[:, 'DBS1':]\n",
    "            mean_df += slice_read_df\n",
    "            \n",
    "        dbs_df_averages = mean_df/len(file_paths) \n",
    "        mut_subtypes = og_df.loc[:,'Mutation Type'].to_frame() \n",
    "        average_freq_df = pd.concat([mut_subtypes, dbs_df_averages], axis=1)\n",
    "        \n",
    "        return average_freq_df\n",
    "\n",
    "    if mut_type == \"ID\":\n",
    "        \n",
    "        outlier_list = outlier_detection(cancer_type, num_data, std_outlier)\n",
    "        outlier_paths = [abs_path(\"ID_\" + cancer_type + \"_\" + outlier + \".csv\", \"File\") for outlier in outlier_list]\n",
    "        file_paths = glob.glob(str(id_freq_folder_path) + \"/ID_\" + cancer_type + '_*' + '.csv')\n",
    "        filtered_file_paths = [x for x in file_paths if x not in outlier_paths]\n",
    "        \n",
    "        #print(\"Dropped \" + str(len(file_paths)-len(filtered_file_paths)) + \" outliers!\")\n",
    "        #print(\"\")\n",
    "        #print(outlier_list)\n",
    "        \n",
    "        og_df = pd.read_csv(filtered_file_paths[0], index_col=0)\n",
    "        mean_df = og_df.loc[:, 'ID1':]\n",
    "        \n",
    "        for file_index in range(1, len(file_paths)):\n",
    "            raw_mean = pd.read_csv(file_paths[file_index])\n",
    "            slice_read_df = raw_mean.loc[:, 'ID1':]\n",
    "            mean_df += slice_read_df\n",
    "            \n",
    "        id_df_averages = mean_df/len(file_paths) \n",
    "        mut_subtypes = og_df.loc[:,'Mutation Type':'ID_type']\n",
    "        average_freq_df = pd.concat([mut_subtypes, id_df_averages], axis=1)\n",
    "        \n",
    "        return average_freq_df\n",
    "  \n",
    "    if mut_type not in [\"SBS\", \"DBS\", \"ID\"]:\n",
    "        raise ValueError(\"It looks like you did not input an appropriate mut_type argument (SBS, DBS, ID)\")\n",
    "\n",
    "    else: \n",
    "        print(\"Completed calculations of average \" + str(mut_type) + \"mutation expected frequency for \" + str(cancer_type))\n",
    "\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mut_prob(cancer_type, prop_data, num_data, freq_folder_path, mut_type, std_outlier):\n",
    "    \"\"\"\n",
    "Output dictionary of probability of mutation for any given k-mer in whole genome sequence\n",
    "\n",
    "Arguments:\n",
    "    cancer_type (str): Cancer type\n",
    "    prop_data (dataframe): Signature proportion data\n",
    "    num_data (dataframe): Signature frequency data\n",
    "    freq_folder_path (str): Folder with expected mutation frequencies for each sample\n",
    "    mut_type (str): \"SBS\", \"DBS\", \"ID\"\n",
    "    std_outlier (float): Parameter to exclude outliers by degrees of standard deviation \n",
    "    \"\"\"\n",
    "    if mut_type == \"SBS\":\n",
    "        \n",
    "        sbs_mut = mut_mean(cancer_type, num_data, freq_folder_path, mut_type, std_outlier)\n",
    "        sbs_mut['total'] = sbs_mut['SubType'].map(kmer_3_count.set_index('kmer')['count'])\n",
    "        sbs_mut_prob = sbs_mut.loc[:,'SBS1':'SBS60'].div(sbs_mut.total, axis=0)\n",
    "        sbs_index_mut_prob = pd.concat([sbs_mut[['Type', 'SubType']], sbs_mut_prob], axis=1)\n",
    "        sbs_index_mut_prob.sort_values(['SubType','Type']).set_index(['SubType','Type'],inplace=True)\n",
    "    \n",
    "        return sbs_index_mut_prob\n",
    "    \n",
    "    if mut_type == \"DBS\":\n",
    "\n",
    "        dbs_mut = mut_mean(cancer_type, num_data, freq_folder_path, mut_type, std_outlier)\n",
    "        dbs_mut['2mer_index'] = dbs_mut['Mutation Type'].apply(lambda x: x[:2])\n",
    "        dbs_mut['total'] = dbs_mut['2mer_index'].map(kmer_2_count.set_index('kmer')['count'])\n",
    "        dbs_mut_prob = dbs_mut.loc[:,'DBS1':'DBS11'].div(dbs_mut['total'], axis=0)\n",
    "        dbs_index_mut_prob = pd.concat([dbs_mut[['Mutation Type', '2mer_index']], dbs_mut_prob], axis=1)\n",
    "        dbs_index_mut_prob.sort_values(['Mutation Type']).set_index(['Mutation Type'],inplace=True)\n",
    "   \n",
    "        return dbs_index_mut_prob\n",
    "    \n",
    "    if mut_type == \"ID\":\n",
    "\n",
    "        id_mut = mut_mean(cancer_type, num_data, freq_folder_path, mut_type, std_outlier)\n",
    "        id_mut['Index'] = id_prop_data['Index']\n",
    "        \n",
    "        id_kmer_count_1_6 = kmer_1_count.append([kmer_2_count,\n",
    "                                                 kmer_3_count,\n",
    "                                                 kmer_4_count,\n",
    "                                                 kmer_5_count,\n",
    "                                                 kmer_6_count])\n",
    "        \n",
    "        id_mut['total'] = id_mut['Index'].map(id_kmer_count_1_6.set_index('kmer')['count'])\n",
    "        \n",
    "        id_mut_prob = id_mut.loc[:,'ID1':'ID17'].div(id_mut['total'], axis=0)\n",
    "        id_index_mut_prob_1_6 = pd.concat([id_mut.loc[:, 'Mutation Type':'ID_type'], id_mut_prob], axis=1)\n",
    "        id_index_mut_prob_1_6.sort_values(['Mutation Type']).set_index(['Mutation Type'],inplace=True)\n",
    "      \n",
    "        return id_index_mut_prob_1_6\n",
    "    \n",
    "    if mut_type not in [\"SBS\", \"DBS\", \"ID\"]:\n",
    "        raise ValueError(\"It looks like you did not input an appropriate mut_type argument (SBS, DBS, ID)\")\n",
    "   \n",
    "    else:\n",
    "        print(\"Completed calculations of \" + str(mut_type) + \"mutation probabilities for \" + str(cancer_type))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def genome_kmer_count(kmer_length, kmer_count_folder_path = abs_path(\"kmer_ref_count\", \"Directory\")):\n",
    "    \"\"\"\n",
    "Output genome kmer counts (Chr1-22, X and Y) of reference genome GChr38\n",
    "\n",
    "Arguments:\n",
    "    kmer_length: Counts of k-mers in reference sequence at specified length\n",
    "    kmer_count_folder_path: Folder path of kmer counts, default is folder of reference genome kmer counts \n",
    "    \"\"\"\n",
    "    chr_index = [i for i in range(kmer_length + 1 ,23)]\n",
    "    chr_index.append('X')\n",
    "    chr_index.append('Y')\n",
    "    \n",
    "    chr_1 = pd.read_csv(kmer_count_folder_path + '/' + str(kmer_length) + '-mer/' + str(kmer_length) + '-mer_chr1.csv')\n",
    "    total_kmer_count = pd.read_csv(kmer_count_folder_path + '/' + str(kmer_length) + '-mer/' + str(kmer_length) + '-mer_chr1.csv').loc[:, 'count']\n",
    "    \n",
    "    for file_name in chr_index:\n",
    "        read_in = pd.read_csv(kmer_count_folder_path + '/' + str(kmer_length) + '-mer/' + str(kmer_length) + '-mer_chr' + str(file_name) + '.csv').loc[:, 'count']\n",
    "        total_kmer_count += read_in\n",
    " \n",
    "    return pd.concat([chr_1.iloc[:,1], total_kmer_count], axis=1)\n",
    "\n",
    "kmer_count_folder_path = abs_path(\"kmer_ref_count\", \"Directory\")\n",
    "#/Users/DavidChen/Desktop/Project/kmer_ref_count\n",
    "\n",
    "kmer_1_count = genome_kmer_count(1, kmer_count_folder_path)\n",
    "kmer_1_count.columns = ['kmer', 'count']\n",
    "kmer_2_count = genome_kmer_count(2, kmer_count_folder_path)\n",
    "kmer_2_count.columns = ['kmer', 'count']\n",
    "kmer_3_count = genome_kmer_count(3, kmer_count_folder_path)\n",
    "kmer_3_count.columns = ['kmer', 'count']\n",
    "kmer_4_count = genome_kmer_count(4, kmer_count_folder_path)\n",
    "kmer_4_count.columns = ['kmer', 'count']\n",
    "kmer_5_count = genome_kmer_count(5, kmer_count_folder_path)\n",
    "kmer_5_count.columns = ['kmer', 'count']\n",
    "kmer_6_count = genome_kmer_count(6, kmer_count_folder_path)\n",
    "kmer_6_count.columns = ['kmer', 'count']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Reference Genome Length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "kmer_ref = (glob.glob(\"/Users/davidchen/Documents/GitHub/SomaticSiMu/kmer_ref_count/6-mer/6-mer_chr*\"))\n",
    "kmer_count = pd.read_csv(kmer_ref[0], index_col=0)['count'].fillna(0)\n",
    "for i in kmer_ref[1:-1]:\n",
    "    sample = pd.read_csv(i, index_col=0)['count'].fillna(0)\n",
    "    kmer_count = kmer_count + sample\n",
    "\n",
    "kmer_reference_count_dict = dict(zip(pd.read_csv(kmer_ref[0], index_col=0)[\"6\"], kmer_count))\n",
    "ref_genome_length = sum(kmer_reference_count_dict.values())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Normalize genome mutation burden to input seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_ref_burden(cancer_type, num_data, std_outlier, input_seq_length):\n",
    "    #outliers = outlier_detection(cancer_type, num_data, std_outlier)\n",
    "    bar = num_data[num_data['Cancer Types'] == cancer_type]\n",
    "    #bar = bar[~bar['Sample Names'].isin(outliers)]\n",
    "    ref_mut_burden = list(bar.iloc[:, 3:].sum(axis=1))\n",
    "    ref_mut_burden = [i * (input_seq_length/ref_genome_length) for i in ref_mut_burden]\n",
    "    return ref_mut_burden"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "ref_mut_burden = normalize_ref_burden(cancer_type=\"Liver-HCC\", num_data=sbs_num_data, std_outlier=3, input_seq_length = 50818467)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[285.4872567904862,\n",
       " 220.3733966099499,\n",
       " 385.6664184019329,\n",
       " 279.553661136415,\n",
       " 161.85049836586006,\n",
       " 75.42413134621097,\n",
       " 299.6206318615537,\n",
       " 157.30083172731568,\n",
       " 154.44647813279164,\n",
       " 166.93643749792108,\n",
       " 120.28073056197361,\n",
       " 216.51569447916893,\n",
       " 132.96098016673798,\n",
       " 371.34275309123046,\n",
       " 290.62509326062946,\n",
       " 196.63901399366517,\n",
       " 177.76568204441836,\n",
       " 229.62842190128544,\n",
       " 213.02127371496374,\n",
       " 275.7997536817985,\n",
       " 330.6033426966601,\n",
       " 351.4833717184208,\n",
       " 141.59323740108644,\n",
       " 184.37394309355886,\n",
       " 149.74111948000052,\n",
       " 185.3599925171217,\n",
       " 246.87563725728828,\n",
       " 160.17248443453383,\n",
       " 309.51572432257035,\n",
       " 151.22884317169184,\n",
       " 200.42751967366982,\n",
       " 325.7422920296222,\n",
       " 42.763406579778334,\n",
       " 174.09827015327232,\n",
       " 310.3806799572746,\n",
       " 306.38458492494095,\n",
       " 198.87059953120215,\n",
       " 108.41353925383123,\n",
       " 123.70595487540245,\n",
       " 182.93811673994978,\n",
       " 113.17079524470464,\n",
       " 257.96436849419683,\n",
       " 380.5804792698719,\n",
       " 282.20042537861,\n",
       " 488.26745579055154,\n",
       " 276.33602617531517,\n",
       " 253.98557257455724,\n",
       " 106.26844927976468,\n",
       " 308.13179530704355,\n",
       " 227.5698274906893,\n",
       " 205.6518517072835,\n",
       " 119.57146694151612,\n",
       " 48.03963595147429,\n",
       " 210.92808107897946,\n",
       " 109.29579400122958,\n",
       " 253.39740274295835,\n",
       " 298.2021046206387,\n",
       " 207.5201558782447,\n",
       " 183.35329544460782,\n",
       " 321.4867103068772,\n",
       " 145.45093953186742,\n",
       " 114.67581804909004,\n",
       " 40.895102408817145,\n",
       " 129.2243718248156,\n",
       " 275.6094634421636,\n",
       " 532.5704834001035,\n",
       " 310.00009947800476,\n",
       " 209.19816980957094,\n",
       " 102.72213117747724,\n",
       " 95.352709169797,\n",
       " 183.9414652762067,\n",
       " 102.37614892359554,\n",
       " 119.5887660542102,\n",
       " 210.3226121346865,\n",
       " 333.3884998404078,\n",
       " 130.67749729111875,\n",
       " 271.7690604240767,\n",
       " 203.14348036664117,\n",
       " 151.29803962246817,\n",
       " 109.76287004396987,\n",
       " 111.97715646881277,\n",
       " 90.76844430586445,\n",
       " 229.61112278859136,\n",
       " 108.32704369036081,\n",
       " 247.7751911173807,\n",
       " 180.58543741355422,\n",
       " 226.03020646091574,\n",
       " 306.3499866995528,\n",
       " 208.9040848937715,\n",
       " 176.00117254962166,\n",
       " 164.61835639691367,\n",
       " 91.42581058823968,\n",
       " 324.73894349336524,\n",
       " 406.3215589586705,\n",
       " 302.23279787836054,\n",
       " 181.95206731638694,\n",
       " 150.9174591431983,\n",
       " 177.19481132551354,\n",
       " 153.66801806155783,\n",
       " 237.86279954366995,\n",
       " 113.94925531593846,\n",
       " 126.87169249842002,\n",
       " 138.35830332729253,\n",
       " 101.3728003873386,\n",
       " 314.8438510323486,\n",
       " 351.98504598654927,\n",
       " 206.43031177851734,\n",
       " 137.87392817185813,\n",
       " 179.5647897646032,\n",
       " 150.65797245278702,\n",
       " 213.0904701657401,\n",
       " 128.48050997896993,\n",
       " 197.9018492203334,\n",
       " 154.35998256932123,\n",
       " 72.31029106127566,\n",
       " 179.4609950884387,\n",
       " 306.7132680661286,\n",
       " 308.9967509417478,\n",
       " 87.42971555590603,\n",
       " 139.81142879359567,\n",
       " 214.54359563204324,\n",
       " 763.2541511757281,\n",
       " 276.11113771029204,\n",
       " 215.23556013980664,\n",
       " 197.91914833302747,\n",
       " 249.81648641528272,\n",
       " 162.5251637609294,\n",
       " 283.7400464083836,\n",
       " 294.93257232145663,\n",
       " 202.1574309430783,\n",
       " 267.3750857997791,\n",
       " 219.88902145451553,\n",
       " 258.1027613957495,\n",
       " 272.51292226992234,\n",
       " 242.15297949180305,\n",
       " 278.9827904175102,\n",
       " 184.96211292515775,\n",
       " 231.5313242976348,\n",
       " 475.0855319176587,\n",
       " 261.3549945822375,\n",
       " 101.45929595080904,\n",
       " 270.9560021274547,\n",
       " 230.78746245178914,\n",
       " 137.89122728455223,\n",
       " 124.58820962280079,\n",
       " 161.26232853426117,\n",
       " 35.09989965629864,\n",
       " 137.04357076254206,\n",
       " 240.0078895177365,\n",
       " 387.75961103791724,\n",
       " 199.33767557394245,\n",
       " 229.74951569014402,\n",
       " 327.52410063711295,\n",
       " 69.57703125561022,\n",
       " 184.3047466427825,\n",
       " 119.39847581457526,\n",
       " 129.6568496421677,\n",
       " 325.9844796073394,\n",
       " 197.1752864871818,\n",
       " 136.5245973817195,\n",
       " 152.40518283488962,\n",
       " 166.106080088605,\n",
       " 356.5866099631759,\n",
       " 252.9822240383003,\n",
       " 203.69705197285188,\n",
       " 127.01008539997271,\n",
       " 153.85830830119275,\n",
       " 132.1133236447278,\n",
       " 180.87952232935368,\n",
       " 93.96878015427019,\n",
       " 107.8772667603146,\n",
       " 135.76343642317977,\n",
       " 146.07370758885446,\n",
       " 298.9286673537903,\n",
       " 64.23160543313793,\n",
       " 186.48443484223722,\n",
       " 84.74835308832284,\n",
       " 126.0759333144921,\n",
       " 177.62728914286566,\n",
       " 1101.7112910355029,\n",
       " 406.9789252410458,\n",
       " 97.30750890422861,\n",
       " 237.88009865636403,\n",
       " 138.1507139749635,\n",
       " 101.52849240158537,\n",
       " 90.07647979810105,\n",
       " 176.64123971930283,\n",
       " 232.2232888053982,\n",
       " 98.31085744048555,\n",
       " 54.90738369102606,\n",
       " 194.13064265302285,\n",
       " 277.6853569654538,\n",
       " 369.7858329487628,\n",
       " 185.51568453136846,\n",
       " 287.8053378914936,\n",
       " 196.1546388382308,\n",
       " 109.34769133931184,\n",
       " 232.89795420046752,\n",
       " 60.47769797852146,\n",
       " 99.02012106094304,\n",
       " 130.97158220691819,\n",
       " 255.26570691391953,\n",
       " 234.47217345562927,\n",
       " 84.76565220101692,\n",
       " 335.7930765048856,\n",
       " 39.511173393290335,\n",
       " 121.68195869019449,\n",
       " 187.28019402616513,\n",
       " 261.9950617519186,\n",
       " 267.09829999667375,\n",
       " 164.72215107307818,\n",
       " 288.4454050611747,\n",
       " 134.27571273148845,\n",
       " 47.52066257065173,\n",
       " 153.32203580767612,\n",
       " 440.6602976564294,\n",
       " 164.61835639691367,\n",
       " 170.81143874139613,\n",
       " 623.5811152836851,\n",
       " 179.2015083980274,\n",
       " 138.68698646848014,\n",
       " 139.81142879359567,\n",
       " 207.38176297669202,\n",
       " 184.73722446013463,\n",
       " 113.62057217475085,\n",
       " 63.90292229195031,\n",
       " 250.681442049987,\n",
       " 318.3209726838597,\n",
       " 133.5145517729487,\n",
       " 199.77015339129457,\n",
       " 328.23336425757043,\n",
       " 221.51513804775954,\n",
       " 66.54968653414532,\n",
       " 151.54022720018537,\n",
       " 60.61609088007414,\n",
       " 137.77013349569364,\n",
       " 266.64852306662755,\n",
       " 57.77903639824419,\n",
       " 285.98893105861464,\n",
       " 211.74113937560145,\n",
       " 270.67921632434934,\n",
       " 229.99170326786123,\n",
       " 188.21434611164574,\n",
       " 516.9147864119565,\n",
       " 206.2400215388824,\n",
       " 218.9894675944231,\n",
       " 180.79302676588324,\n",
       " 262.6351289215998,\n",
       " 157.94089889699683,\n",
       " 170.2924653605736,\n",
       " 154.6367683724266,\n",
       " 230.78746245178914,\n",
       " 141.24725514720473,\n",
       " 152.83766065224174,\n",
       " 100.64623765418703,\n",
       " 217.06926608537967,\n",
       " 81.46152167644667,\n",
       " 238.45096937526884,\n",
       " 131.38676091157623,\n",
       " 376.6535806883146,\n",
       " 189.25229287329086,\n",
       " 365.08047429597167,\n",
       " 348.19654030654465,\n",
       " 122.73720456453368,\n",
       " 112.08095114497728,\n",
       " 361.65524998254284,\n",
       " 268.03245208215435,\n",
       " 238.39907203718658,\n",
       " 174.15016749135455,\n",
       " 313.8751007214798,\n",
       " 204.92528897413194,\n",
       " 293.02966992510727,\n",
       " 92.74054315299014,\n",
       " 97.51509825655764,\n",
       " 306.3326875868587,\n",
       " 334.58213861629963,\n",
       " 319.8259954882451,\n",
       " 136.21321335322597,\n",
       " 115.886755937676,\n",
       " 186.7439215326485,\n",
       " 200.7908010402456,\n",
       " 160.81255160421497,\n",
       " 132.14792187011597,\n",
       " 181.83097352752836,\n",
       " 145.43364041917332,\n",
       " 238.08768800869305,\n",
       " 213.62674265925673,\n",
       " 12.022883322389134,\n",
       " 235.40632554110985,\n",
       " 189.0966008590441,\n",
       " 356.93259221705756,\n",
       " 163.13063270522235,\n",
       " 182.3326477956568,\n",
       " 142.6657823881197,\n",
       " 290.14071810519505,\n",
       " 100.71543410496338,\n",
       " 167.17862507563828,\n",
       " 275.695959005634,\n",
       " 310.2076888303338,\n",
       " 69.80191972063332,\n",
       " 169.27181771162256,\n",
       " 144.91466703835076,\n",
       " 325.55200178998723,\n",
       " 137.42415124181193,\n",
       " 278.4811161493817,\n",
       " 148.72047183104948,\n",
       " 227.3795372510544,\n",
       " 58.83428227258338,\n",
       " 323.54530471747336,\n",
       " 237.17083503590655,\n",
       " 82.74165601580897,\n",
       " 202.24392650654875,\n",
       " 167.939786034178,\n",
       " 160.58766313919185,\n",
       " 186.20764903913187,\n",
       " 127.70204990773611,\n",
       " 385.28583792266306,\n",
       " 85.35382203261581,\n",
       " 160.24168088531016,\n",
       " 210.44370592354508,\n",
       " 551.0632348700805,\n",
       " 92.2215697721676,\n",
       " 396.30537270879523,\n",
       " 252.44595154478367,\n",
       " 231.8773065515165,\n",
       " 260.2997487078983]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_mut_burden"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "1\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "10\n",
      "11\n",
      "12\n",
      "13\n",
      "14\n",
      "15\n",
      "16\n",
      "17\n",
      "18\n",
      "19\n",
      "20\n",
      "21\n",
      "22\n",
      "23\n",
      "24\n",
      "25\n",
      "26\n",
      "27\n",
      "28\n",
      "29\n",
      "30\n",
      "31\n",
      "32\n",
      "33\n",
      "34\n",
      "35\n",
      "36\n",
      "37\n",
      "38\n",
      "39\n",
      "40\n",
      "41\n",
      "42\n",
      "43\n",
      "44\n",
      "45\n",
      "46\n",
      "47\n",
      "48\n",
      "49\n",
      "50\n",
      "51\n",
      "52\n",
      "53\n",
      "54\n",
      "55\n",
      "56\n",
      "57\n",
      "58\n",
      "59\n",
      "60\n",
      "61\n",
      "62\n",
      "63\n",
      "64\n",
      "65\n",
      "66\n",
      "67\n",
      "68\n",
      "69\n",
      "70\n",
      "71\n",
      "72\n",
      "73\n",
      "74\n",
      "75\n",
      "76\n",
      "77\n",
      "78\n",
      "79\n",
      "80\n",
      "81\n",
      "82\n",
      "83\n",
      "84\n",
      "85\n",
      "86\n",
      "87\n",
      "88\n",
      "89\n",
      "90\n",
      "91\n",
      "92\n",
      "93\n",
      "94\n",
      "95\n",
      "96\n",
      "97\n",
      "98\n",
      "99\n",
      "100\n",
      "101\n",
      "102\n",
      "103\n",
      "104\n",
      "105\n",
      "106\n",
      "107\n",
      "108\n",
      "109\n",
      "110\n",
      "111\n",
      "112\n",
      "113\n",
      "114\n",
      "115\n",
      "116\n",
      "117\n",
      "118\n",
      "119\n",
      "120\n",
      "121\n",
      "122\n",
      "123\n",
      "124\n",
      "125\n",
      "126\n",
      "127\n",
      "128\n",
      "129\n",
      "130\n",
      "131\n",
      "132\n",
      "133\n",
      "134\n",
      "135\n",
      "136\n",
      "137\n",
      "138\n",
      "139\n",
      "140\n",
      "141\n",
      "142\n",
      "143\n",
      "144\n",
      "145\n",
      "146\n",
      "147\n",
      "148\n",
      "149\n",
      "150\n",
      "151\n",
      "152\n",
      "153\n",
      "154\n",
      "155\n",
      "156\n",
      "157\n",
      "158\n",
      "159\n",
      "160\n",
      "161\n",
      "162\n",
      "163\n",
      "164\n",
      "165\n",
      "166\n",
      "167\n",
      "168\n",
      "169\n",
      "170\n",
      "171\n",
      "172\n",
      "173\n",
      "174\n",
      "175\n",
      "176\n",
      "177\n",
      "178\n",
      "179\n",
      "180\n",
      "181\n",
      "182\n",
      "183\n",
      "184\n",
      "185\n",
      "186\n",
      "187\n",
      "188\n",
      "189\n",
      "190\n",
      "191\n",
      "192\n",
      "193\n",
      "194\n",
      "195\n",
      "196\n",
      "197\n",
      "198\n",
      "199\n",
      "200\n",
      "201\n",
      "202\n",
      "203\n",
      "204\n",
      "205\n",
      "206\n",
      "207\n",
      "208\n",
      "209\n",
      "210\n",
      "211\n",
      "212\n",
      "213\n",
      "214\n",
      "215\n",
      "216\n",
      "217\n",
      "218\n",
      "219\n",
      "220\n",
      "221\n",
      "222\n",
      "223\n",
      "224\n",
      "225\n",
      "226\n",
      "227\n",
      "228\n",
      "229\n",
      "230\n",
      "231\n",
      "232\n",
      "233\n",
      "234\n",
      "235\n",
      "236\n",
      "237\n",
      "238\n",
      "239\n",
      "240\n",
      "241\n",
      "242\n",
      "243\n",
      "244\n",
      "245\n",
      "246\n",
      "247\n",
      "248\n",
      "249\n",
      "250\n",
      "251\n",
      "252\n",
      "253\n",
      "254\n",
      "255\n",
      "256\n",
      "257\n",
      "258\n",
      "259\n",
      "260\n",
      "261\n",
      "262\n",
      "263\n",
      "264\n",
      "265\n",
      "266\n",
      "267\n",
      "268\n",
      "269\n",
      "270\n",
      "271\n",
      "272\n",
      "273\n",
      "274\n",
      "275\n",
      "276\n",
      "277\n",
      "278\n",
      "279\n",
      "280\n",
      "281\n",
      "282\n",
      "283\n",
      "284\n",
      "285\n",
      "286\n",
      "287\n",
      "288\n",
      "289\n",
      "290\n",
      "291\n",
      "292\n",
      "293\n",
      "294\n",
      "295\n",
      "296\n",
      "297\n",
      "298\n",
      "299\n",
      "300\n",
      "301\n",
      "302\n",
      "303\n",
      "304\n",
      "305\n",
      "306\n",
      "307\n",
      "308\n",
      "309\n",
      "310\n",
      "311\n",
      "312\n",
      "313\n",
      "314\n",
      "315\n",
      "316\n",
      "317\n",
      "318\n",
      "319\n",
      "320\n",
      "321\n",
      "322\n",
      "323\n",
      "324\n",
      "325\n",
      "326\n",
      "327\n",
      "328\n",
      "329\n",
      "330\n",
      "331\n",
      "332\n",
      "333\n",
      "334\n",
      "335\n",
      "336\n",
      "337\n",
      "338\n",
      "339\n",
      "340\n",
      "341\n",
      "342\n",
      "343\n",
      "344\n",
      "345\n",
      "346\n",
      "347\n",
      "348\n",
      "349\n",
      "350\n",
      "351\n",
      "352\n",
      "353\n",
      "354\n",
      "355\n",
      "356\n",
      "357\n",
      "358\n",
      "359\n",
      "360\n",
      "361\n",
      "362\n",
      "363\n",
      "364\n",
      "365\n",
      "366\n",
      "367\n",
      "368\n",
      "369\n",
      "370\n",
      "371\n",
      "372\n",
      "373\n",
      "374\n",
      "375\n",
      "376\n",
      "377\n",
      "378\n",
      "379\n",
      "380\n",
      "381\n",
      "382\n",
      "383\n",
      "384\n",
      "385\n",
      "386\n",
      "387\n",
      "388\n",
      "389\n",
      "390\n",
      "391\n",
      "392\n",
      "393\n",
      "394\n",
      "395\n",
      "396\n",
      "397\n",
      "398\n",
      "399\n",
      "400\n",
      "401\n",
      "402\n",
      "403\n",
      "404\n",
      "405\n",
      "406\n",
      "407\n",
      "408\n",
      "409\n",
      "410\n",
      "411\n",
      "412\n",
      "413\n",
      "414\n",
      "415\n",
      "416\n",
      "417\n",
      "418\n",
      "419\n",
      "420\n",
      "421\n",
      "422\n",
      "423\n",
      "424\n",
      "425\n",
      "426\n",
      "427\n",
      "428\n",
      "429\n",
      "430\n",
      "431\n",
      "432\n",
      "433\n",
      "434\n",
      "435\n",
      "436\n",
      "437\n",
      "438\n",
      "439\n",
      "440\n",
      "441\n",
      "442\n",
      "443\n",
      "444\n",
      "445\n",
      "446\n",
      "447\n",
      "448\n",
      "449\n",
      "450\n",
      "451\n",
      "452\n",
      "453\n",
      "454\n",
      "455\n",
      "456\n",
      "457\n",
      "458\n",
      "459\n",
      "460\n",
      "461\n",
      "462\n",
      "463\n",
      "464\n",
      "465\n",
      "466\n",
      "467\n",
      "468\n",
      "469\n",
      "470\n",
      "471\n",
      "472\n",
      "473\n",
      "474\n",
      "475\n",
      "476\n",
      "477\n",
      "478\n",
      "479\n",
      "480\n",
      "481\n",
      "482\n",
      "483\n",
      "484\n",
      "485\n",
      "486\n",
      "487\n",
      "488\n",
      "489\n",
      "490\n",
      "491\n",
      "492\n",
      "493\n",
      "494\n",
      "495\n",
      "496\n",
      "497\n",
      "498\n",
      "499\n"
     ]
    }
   ],
   "source": [
    "simulated_mut_burden = mut_burden(\"Liver-HCC\", \"End\", 0,499)['sbs'].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "means = [np.mean(ref_mut_burden) , np.mean(simulated_mut_burden)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "yerr = [np.std(ref_mut_burden) , np.std(simulated_mut_burden)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot mean mutation burden "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAagAAAEYCAYAAAAJeGK1AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAAgAElEQVR4nO3debgcVZnH8e+PhH2RJXgJYYkoUQOOEa6A4nIREGQUUAFBZdfojLgyo4gMBoUZVDTqMI7GUQmKCMoioiwBaTcWTSAsAcEAQWJCIhggbIGEd/4454ZKp5dKcvt2Jff3eZ773K5Tp6rerq6qt+rU6WpFBGZmZlWzVrcDMDMza8QJyszMKskJyszMKskJyszMKskJyszMKskJyszMKskJagBJep+kq7sdRz9J60v6haTHJP20A/M/XdLDkh4a6HmvjiTNkrTPSky3naQnJA3rQEzHSPr9QM+3ybJGSwpJwwdjeasir+8duh3Hysjr+GXdjmMwVDJBSXqvpKl5I5or6QpJb+h2XO1ExHkR8dZux1FwCNADbBERh9aPlDRB0nN5PT8q6XpJryszY0nbAicCYyNiq4ENe80jaRtJF+WE/pik2yUdAxARf42IjSJiSZdjnCDpRx2c/yxJz0oaUVc+PR90R5eczwodoCXVJH2gWJbX931l57ECy5ol6em8Ty2Q9Mu8r9hKqFyCkvQp4OvAf5IOrtsB3wIO6mZc7VT0rHF74J6IWNyizgURsREwArgOKHultT3wSETMX9GglFRu2+uwHwIPktbbFsBRwLyuRtQd9wNH9A9IehWwfvfC6Yh35H1qJOkz/u+VmUlFjymDKyIq8we8CHgCOLRFnXVJCWxO/vs6sG4e1wfMBj4NzAfmAgcDBwD3AP8ATi7MawLwM+ACYCFwM/DqwviTgHvzuDuBdxbGHQP8AZiY53t6Lvt9Hq88bj7wGHAbsHPhfZ4L/B14ADgFWKsw398DZwELSDv021qsj1cCNeBRYAZwYC4/DXgWeC6v0+MbTDsB+FFheCwQwJaFOL+X1+Pf8nscBuwDPA08n+d9Tq6/B3B9juVWoK8w7xpwRl5nTwMvA14BTMnr727gsEL9c4D/AX6Z1/9NwEsL43cqTDuv/3MlnXT1f26PABcCmzdZd5sBl+fPYUF+vU1dzF/MMS8ErgZGFMYfmT+/R4DPAbOAfZos6wlgXJNxo/N6H15Y7ul5XT4B/IKU1M4DHgf+BIxuNG1h+g8Ut6fCuG+QEuXjwDTgjbl8/7rt5dZW20AeN4y0nT4M3Ad8pD6Wuvc5i7St/6lQdlZed1F4T0vjr38PwG9z3SdznO9p9TmStrklwDO5/tm5PICXdWB/XGYbIB97Gn02TT6fyOvxL8D9uezf8/qfAxxXF/u6Oba/kvaDbwPr1x0PT+SF4+Gxg31cX5W/rgdQ9+HuDyxutoHnOl8AbgReDGxJ2om/WPhAFgOnAmsDH8wb3Y+BjUkHtWeAHXL9CaQd8pBc/9/yBrh2Hn8osDXpoPeevFOMLGxYi4GPAsNJZ4HFHWk/0gFgU1KyemVh2nOBn+eYRpOS5/GF+T6XYx8G/EveMNVgXawNzAROBtYB3kI6kL688P5+1GJdLh2fpz+TdLDpP1BeCnwH2DCv7z8CHypu/IV5jSIdqA/I62vfPNyf7GqknWinvL5eRDpQHpuHd8nL3inXP4eUfHbL488DfpLHbUza2U4E1svDu+dxnyBtH9uQdt7vAOc3ef9bAO8GNsjz+Clwad3B5F5gTP58a8CZedxY0gHvTXk5X8vbQ7MEdQ0p0R0ObFc3bjTLJ6iZwEvzerozbyP75HVxLvCDRtMWpm+WoN6f3/fwvP4eAtZrtr202QY+DPwZ2BbYnHQF3i5B7UM6GXklafvuv6oslaDy8NID9Ap8jh+oi6V4kB+Q/bH4HvPrDYDJwLnNYmny3qbk9bk+6Zg4D9g5fwY/rov968Bluf7GpJOZ/6o7Hn6BdKw4AHgK2Kzbx/qyf10PoO7DfR/wUJs69wIHFIb3A2YVPpCneeEMb+P8Ye5eqD8NOLiwQ95YGLcW6cD3xibLng4cVNiw/lo3funGRkoW95CuKtYq1BkGLCLdu+kv+xBQK8xjZmHcBvk9bNUgnjeSDjDF+Z8PTCi8v3YJ6lnSFc8SUkLpy+N6cpzrF+ofAVxXWNfFBPUZ4Id1878KODq/rgFfKIx7D/C7uvrfAT6fX58D/F9h3AHAnwtx3NLkPd0F7F0YHkk6wDQ96SnUHQcsKAzXgFMKw/8KXJlfn0pOmHl4w7wumyWozUgnADPyup4OvDaPG83yCepzhWm/ClxRGH4HML3RtIXpGyaoBnEtILca1G8vJbaBXwMfLox7a30sdcuaRUpQpwD/RTr4TiEly5VOUCU/x4YJigHcHwvv8QnSPrWYlMxe1SyWJu/tLYXh75NPivLwmELsIp00F1sWXscLV159pONhcduYD+zRbl+oyl/V7gM8Aoxo0/a6NekyvN8DuWzpPOKFm81P5//Ftv6ngY0Kww/2v4iI50mXxFsDSDoq38B9VNKjpLOYEY2mrRcRvwbOJjVTzZM0SdImefp1GryHUYXhhwrzeSq/LMbcb2vgwRx3s3m1c2FEbEo6GN0B7JrLtyeddc0tvP/vkM6iG9keOLS/bq7/BlKC6PdgXf3d6+q/Dyh2uCj2DnyKF9bBtqQTlWZxXFKY512khNBTX1HSBpK+I+kBSY+Tmo82retN1yyGrVl223mStP02FBELIuKkiNgpxzIduFSSmkxSv8222oZLk3SipLtyR41HSVdoI5pUb7cNLLMOWHabbuWHwHtJB+dzV/AtLKfk59jMQO6P/Q7O+9S6wAnAbyStSEei4jpttY63JCXMaYXP58pc3u+RWPYedHEbrryqJagbSE1wB7eoM4e04/TbLpetrKU9bPKN+22AOZK2B75L2sC2yBvcHaSzln7RasYR8c2I2JXUrDWG1Jb8MOmMvv49/G0lYp8DbFvX4WCl5hURD5POHCdIGknaKRaR7rlsmv82yQfYRh4kXUFtWvjbMCLOLC6mrv5v6upvFBH/UiLcB0nNX83Gva1uvutFRKN1ciLwctIV9iak5jpY9jNuZi7LbjsbkJqa2srr+izSwWfzMtO08GT+v0GhrOHBUNIbSVe6h5GaeTYl3R/tf7/123O7bWCZdUDa9tqKiAdITekHABc3eU9t309Bu8+x1X46kPvjMiJiSURcTDpB6u+FXOa9FeNttY4fJp2s7FT4fF4UqYPGGqFSCSoiHiM1nfyPpIPzmdHakt4m6cu52vnAKZK2zN1VTwVWpWvsrpLela/aPkHaIW8kNdkE6R4Wko4lXUGVIum1knaXtDZpo3wGWJKv7i4EzpC0cU6En1rJ93BTnven83rqIzX//GQl5kVE/JnULPfpiJhL6hTwVUmbSFpL0kslvbnJ5D8C3iFpP0nDJK0nqU/SNk3qXw6MkXRkjn3tvM5eWSLUy4GtJH1C0rp5Pe6ex32btG63B8jbSbMeoBuTdvBHJW0OfL7Esvv9DHi7pDdIWofUzt90f5L0JUk7SxouaWPSvYyZEdH0qquMiPg76WD6/rzej6N58t6Y1Oz0d2C4pFOBTQrj5wGj+094SmwDFwIfU+pCvxmpc0pZx5Oasp5sMG468K68/78s1y2aBxS/w9Tuc6yvv9QA74/LyL1VDyI1796Vi9u9t3oXAsdIGptPgpa+t9xy8l1goqQX52WOkrTfqsZeFZVKUAAR8TXSBnIKaUd6kHQVc2mucjowldQr7nZSz7vTV2GRPyfdD1lA6pX1roh4LiLuJLX930DawF9Fusld1iakjWcBL/T0OiuP+ygpsdxH6iH0Y1Jb8wqJiGeBA4G3kc6mvgUclRPNyvoKMD5v8EeRmj/uzO/jZyzbZFeM5UHSVwFO5oXP7d9pso1FxELSPYvDSVeCDwFfIjWLtJSn3ZeUjB8i9XjaK4/+Bumm8dWSFpJONnZvNB/SDeb1SevuRlLzSCkRMYPU2+rHpLPcBaTm4WY2AC4h3Zu4j3TGfmDZ5bXxQdK6foR0tX59k3pXAVeQ7o0+QDppKjYf9X/F4BFJN+fXrbaB7+Z53kraDxtdDTUUEfdGxNQmoyeS7ufNI3UyOK9u/ARgcm7WOoz2n+M3gEOUvpf0zQbLG5D9seAXkp4g9ZQ8g3QfdkbJ97aMiLiC9P5+Teo48+u6Kp/J5Tfm5s1rSFeTawTlG2dDkqQJpJut7+92LGZmtqzKXUGZmZmBE5SZmVXUkG7iMzOz6vIVlJmZVdJq/TDCESNGxOjRo7sdhplZV02bNu3hiNiyfc3Vy2qdoEaPHs3Uqc16qpqZDQ2Syj7FY7XiJj4zM6skJygzM6skJygzM6skJygzM6skJygzM6skJygzM6skJygzM6skJygzM6skJygzM6skJygzM6uk1fpRR2VMnHJPt0MwGzCf3HdMt0MwGzS+gjIzs0pygjIzs0pygjIzs0pygjIzs0rqWIKStK2k6yTdJWmGpI/n8s0lTZH0l/x/s8I0n5U0U9LdkvbrVGxmZlZ9nbyCWgycGBGvBPYAPiJpLHAScG1E7Ahcm4fJ4w4HdgL2B74laVgH4zMzswrrWIKKiLkRcXN+vRC4CxgFHARMztUmAwfn1wcBP4mIRRFxPzAT2K1T8ZmZWbUNyvegJI0GXgPcBPRExFxISUzSi3O1UcCNhclm57L6eY0HxgP09PRQq9VaLnvUM4tWLXizCqnV5nQ7BLNB0/EEJWkj4CLgExHxuKSmVRuUxXIFEZOASQC9vb3R19fXcvn+oq6tSQ7r8xd1bejoaC8+SWuTktN5EXFxLp4naWQePxKYn8tnA9sWJt8G8OmimdkQ1clefAK+B9wVEV8rjLoMODq/Phr4eaH8cEnrSnoJsCPwx07FZ2Zm1dbJJr49gSOB2yVNz2UnA2cCF0o6HvgrcChARMyQdCFwJ6kH4EciYkkH4zMzswrrWIKKiN/T+L4SwN5NpjkDOKNTMZmZ2erDT5IwM7NKcoIyM7NKcoIyM7NKcoIyM7NKcoIyM7NKcoIyM7NKcoIyM7NKcoIyM7NKcoIyM7NKcoIyM7NKcoIyM7NKcoIyM7NKcoIyM7NKcoIyM7NKcoIyM7NKcoIyM7NKcoIyM7NKcoIyM7NK6liCkvR9SfMl3VEou0DS9Pw3S9L0XD5a0tOFcd/uVFxmZrZ6GN7BeZ8DnA2c218QEe/pfy3pq8Bjhfr3RsS4DsZjZmarkY4lqIj4raTRjcZJEnAY8JZOLd/MzFZvnbyCauWNwLyI+Euh7CWSbgEeB06JiN81mlDSeGA8QE9PD7VareWCRj2zaEACNquCWm1Ot0MwGzTdSlBHAOcXhucC20XEI5J2BS6VtFNEPF4/YURMAiYB9Pb2Rl9fX8sFTZxyz4AFbdZth/WN6XYIZoNm0HvxSRoOvAu4oL8sIhZFxCP59TTgXsB7opnZENaNbub7AH+OiNn9BZK2lDQsv94B2BG4rwuxmZlZRXSym/n5wA3AyyXNlnR8HnU4yzbvAbwJuE3SrcDPgA9HxD86FZuZmVVfJ3vxHdGk/JgGZRcBF3UqFjMzW/34SRJmZlZJTlBmZlZJTlBmZlZJTlBmZlZJTlBmZlZJTlBmZlZJTlBmZlZJTlBmZlZJTlBmZlZJTlBmZlZJbR91JGlPYAKwfa4vICJih86GZmZmQ1mZZ/F9D/gkMA1Y0tlwzMzMkjIJ6rGIuKLjkZiZmRWUSVDXSfoKcDGw9PfTI+LmjkVlZmZDXpkEtXv+31soC+AtAx+OmZlZ0jZBRcRegxGImZlZUdtu5pJ6JH1P0hV5eGzh13HNzMw6osz3oM4BrgK2zsP3AJ/oVEBmZmZQLkGNiIgLgecBImIxJbqbS/q+pPmS7iiUTZD0N0nT898BhXGflTRT0t2S9luJ92JmZmuQMgnqSUlbkDpGIGkP4LES050D7N+gfGJEjMt/v8rzHAscDuyUp/mWpGEllmFmZmuoMr34PgVcBrxU0h+ALYFD2k0UEb+VNLpkHAcBP4mIRcD9kmYCuwE3lJzezMzWMGV68d0s6c3Ay0mPObo7Ip5bhWWeIOkoYCpwYkQsAEYBNxbqzM5ly5E0HhgP0NPTQ61Wa7mwUc8sajnebHVSq83pdghmg6ZpgpL0riajxkgiIi5eieX9L/BFUnPhF4GvAseREl+9aDSDiJgETALo7e2Nvr6+lgucOOWelQjTrJoO6xvT7RDMBk2rK6h35P8vBl4P/DoP7wXUSE+WWCERMa//taTvApfnwdnAtoWq2wA+VTQzG8KadpKIiGMj4ljSlczYiHh3RLyb1JFhpUgaWRh8J9Dfw+8y4HBJ60p6CbAj8MeVXY6Zma3+ynSSGB0RcwvD84C27QySzgf6gBGSZgOfB/okjSMlvVnAhwAiYoakC4E7gcXARyLCT043MxvCyiSomqSrgPNJieVw4Lp2E0XEEQ2Kv9ei/hnAGSXiMTOzIaBML74TJL0TeFMumhQRl3Q2LDMzG+paJihJawG3RcTOgJOSmZkNmpZPkoiI54FbJW03SPGYmZkB5e5BjQRmSPoj8GR/YUQc2LGozMxsyCuToE7reBRmZmZ1ynSS+M1gBGJmZlbUNkFJWsgLjx1aB1gbeDIiNulkYGZmNrSVuYLauDgs6WDSk8bNzMw6pszvQS0jIi4F3tKBWMzMzJYq08RXfKr5WkAvTZ40bmZmNlDK9OJ7R+H1YtIz9A7qSDRmZmZZmXtQxw5GIGZmZkVN70FJWk/S0ZIOVPJpSZdL+oakEYMZpJmZDT2tOkmcC7yV9Iu3NWB74GxgIXBOpwMzM7OhrVUT39iI2FnScGB2RLw5l18p6dZBiM3MzIawVldQzwJExGKW//l1/5igmZl1VKsrqG0kfRNQ4TV5eFTHIzMzsyGtVYL698LrqXXj6oeXI+n7wNuB+fn3pJD0FVK39WeBe4FjI+JRSaOBu4C78+Q3RsSHy7wBMzNbMzVNUBExeRXnfQ6pU8W5hbIpwGcjYrGkLwGfBT6Tx90bEeNWcZlmZraGWOFHHZUVEb8F/lFXdnW+pwVwI7BNp5ZvZmartzJPkuiU44ALCsMvkXQL8DhwSkT8rtFEksYD4wF6enqo1WotFzLqmUUDEqxZFdRq9f2VzNZcXUlQkj5HemzSebloLrBdRDwiaVfgUkk7RcTj9dNGxCRgEkBvb2/09fW1XNbEKfcMZOhmXXVY35huh2A2aMo8LHZL4IPA6GL9iDhuZRYo6WhS54m9IyLyvBYBi/LraZLuBcZQojOGmZmtmcpcQf0c+B1wDav4/SdJ+5M6Rbw5Ip4qlG8J/CMilkjaAdgRuG9VlmVmZqu3Mglqg4j4TPtqy5J0PtAHjJA0G/g8qdfeusAUSfBCd/I3AV+QtJiUBD8cEf9oOGMzMxsSyiSoyyUdEBG/WpEZR8QRDYq/16TuRcBFKzJ/MzNbs5XpZv5xUpJ6RtLC/Ldc5wUzM7OBVOb3oDYejEDMzMyKSnUzl3Qg6T4RQC0iLu9cSGZmZiWa+CSdSWrmuzP/fTyXmZmZdUyZK6gDgHER8TyApMnALcBJnQzMzMyGtrLP4tu08PpFnQjEzMysqMwV1H8Bt0i6jvRbUG8ifZ/JzMysY8r04jtfUg14LSlBfSYiHup0YGZmNrQ1beKT9Ir8fxdgJDAbeBDYOpeZmZl1TKsrqE+Rftbiqw3GBfCWjkRkZmZG61/UHZ9fvi0inimOk7ReR6MyM7Mhr0wvvutLlpmZmQ2YpldQkrYCRgHrS3oNqYMEwCbABoMQm5mZDWGt7kHtBxwDbAN8rVC+EDi5gzGZmZm1vAc1GZgs6d355zDMzMwGTZnvQV0k6Z+BnYD1CuVf6GRgZmY2tJV5WOy3gfcAHyXdhzoU2L7DcZmZ2RBXphff6yPiKGBBRJwGvA7YtrNhmZnZUFcmQT2d/z8laWvgOeAl7SaS9H1J8yXdUSjbXNIUSX/J/zcrjPuspJmS7pa034q+ETMzW7OUSVCXS9oU+ApwMzAL+EmJ6c4B9q8rOwm4NiJ2BK7Nw0gaCxxOus+1P/AtScNKLMPMzNZQZRLUlyPi0dyTb3vgFcDp7SaKiN8C/6grPgiYnF9PBg4ulP8kIhZFxP3ATGC3ErGZmdkaqszPbdwA7AIQEYuARZJu7i9bQT0RMTfPa66kF+fyUcCNhXqzc9lyJI0nPSOQnp4earVaywWOembRSoRpVk212pxuh2A2aKryJAk1KItGFSNiEjAJoLe3N/r6+lrOeOKUe1Y1NrPKOKxvTLdDMBs0g/0kiXmSRuarp5HA/Fw+m2V7Bm4D+FTRzGwIG+wnSVwGHA2cmf//vFD+Y0lfA7YGdgT+OEDLNDOz1VCZe1A7S9qpvrDdkyQknQ/0ASMkzQY+T0pMF0o6Hvgr6Uu/RMQMSRcCdwKLgY9ExJIVeSNmZrZmKZOgnii8Xg94O3BXu4ki4ogmo/ZuUv8M4IwS8ZiZ2RBQ5ll8y/yirqSzSE1yZmZmHVPme1D1NgB2GOhAzMzMitpeQUm6nRe6fA8DtgT8JHMzM+uoMveg3l54vRiYFxGLOxSPmZkZ0PqLupvnlwvrRm0iiYiof4yRmZnZgGl1BfUw6Qu0/VdLxac9BL4PZWZmHdQqQf036XtMfwDOB34fEQ0fP2RmZjbQmvbii4iPA+OAnwJHArdI+rKktr8FZWZmtqpadjOP5Drg08C3gWOBfQYjMDMzG9padZLYkPQ7Te8hdS2/GNglIh4cpNjMzGwIa3UPaj7wF9L9p5mkjhGvlfRagIi4uPPhmZnZUNUqQf2UlJRekf+KgnRFZWZm1hGtfm7jmEGMw8zMbBkr8yw+MzOzjnOCMjOzSnKCMjOzSirzsFgkvR4YXawfEed2KCYzM7NSP7fxQ+ClwHSg/2fYA3CCMjOzjilzBdULjB2o5/BJejlwQaFoB+BUYFPgg8Dfc/nJEfGrgVimmZmtfsokqDuArYC5A7HAiLib9Iw/JA0D/gZcQnqM0sSIOGsglmNmZqu3MglqBHCnpD8Ci/oLI+LAAVj+3sC9EfGApLaVzcxs6CiToCZ0cPmHkx6l1O8ESUcBU4ETI2JB/QSSxgPjAXp6eqjVai0XMOqZRS3Hm61OarU53Q7BbNCoWz/xJGkdYA6wU0TMk9RD+pHEAL4IjIyI41rNo7e3N6ZOndpyOROn3DNAEZt13yf3HdPtEKyCJE2LiN5uxzHQ2n4PStIekv4k6QlJz0paIunxAVj224CbI2IeQETMi4glEfE88F1gtwFYhpmZrabKfFH3bOAI0pPN1wc+kMtW1REUmvckjSyMeyepc4aZmQ1Rpb6oGxEzJQ2LiCXADyRdvyoLlbQBsC/woULxlyWNIzXxzaobZ2ZmQ0yZBPVUvl80XdKXSd3NN1yVhUbEU8AWdWVHrso8zcxszVKmie/IXO8E4ElgW+DdnQzKzMys7RVU/o7S+qRedacNQkxmZmalevG9g/Qcvivz8DhJl3U6MDMzG9rKNPFNIHX5fhQgIqaTnmxuZmbWMWUS1OKIeKzjkZiZmRWUelispPcCwyTtCHwMWKVu5mZmZu2UuYL6KLAT6UGx5wOPA5/oZFBmZmZlevE9BXwu/5mZmQ2KpgmqXU+9Afq5DTMzs4ZaXUG9DniQ1Kx3E+AfbDIzs0HTKkFtRXpe3hHAe4FfAudHxIzBCMzMzIa2pp0k8k9fXBkRRwN7ADOBmqSPDlp0ZmY2ZLXsJCFpXeCfSVdRo4FvAhd3PiwzMxvqWnWSmAzsDFwBnBYR/n0mMzMbNK2uoI4kPb18DPAxaWkfCQEREZt0ODYzMxvCmiaoiCjzJV4zM7OOcBIyM7NKKvWT7wNN0ixgIbCE9DDaXkmbAxeQOmPMAg6LiAXdiM/MzLqvm1dQe0XEuIjozcMnAddGxI7AtXnYzMyGqCo18R0ETM6vJwMHdzEWMzPrsq408QEBXC0pgO9ExCSgJyLmAkTEXEkvbjShpPHAeICenh5qtVrLBY16ZtFAxm3WVbXanG6HYDZoupWg9oyIOTkJTZH057IT5mQ2CaC3tzf6+vpa1p845Z5VidOsUg7rG9PtEMwGTVea+CJiTv4/H7iE9JPy8ySNBMj/53cjNjMzq4ZBT1CSNpS0cf9r4K3AHcBlwNG52tHAzwc7NjMzq45uNPH1AJfkJ1MMB34cEVdK+hNwoaTjgb8Ch3YhNjMzq4hBT1ARcR/w6gbljwB7D3Y8ZmZWTVXqZm5mZraUE5SZmVWSE5SZmVWSE5SZmVWSE5SZmVWSE5SZmVWSE5SZmVWSE5SZmVWSE5SZmVWSE5SZmVWSE5SZmVWSE5SZmVWSE5SZmVWSE5SZmVWSE5SZmVWSE5SZmVWSE5SZmVWSE5SZmVXSoCcoSdtKuk7SXZJmSPp4Lp8g6W+Spue/AwY7NjMzq47hXVjmYuDEiLhZ0sbANElT8riJEXFWF2IyM7OKGfQEFRFzgbn59UJJdwGjBjsOMzOrtm5cQS0laTTwGuAmYE/gBElHAVNJV1kLGkwzHhgP0NPTQ61Wa7mMUc8sGtCYzbqpVpvT7RDMBo0iojsLljYCfgOcEREXS+oBHgYC+CIwMiKOazWP3t7emDp1asvlTJxyzwBFbNZ9n9x3TLdDsAqSNC0iersdx0DrSi8+SWsDFwHnRcTFABExLyKWRMTzwHeB3boRm5mZVUM3evEJ+B5wV0R8rVA+slDtncAdgx2bmZlVRzfuQe0JHAncLml6LjsZOELSOFIT3yzgQ12IzczMKqIbvfh+D6jBqF8NdixmZlZdfpKEmZlVkhOUmZlVkhOUmZlVkhOUmZlVkhOUmZlVkhOUmZlVkhOUmZlVkhOUmZlVkhOUmZlVkhOUmZlVkhOUmZlVkhOUmZlVkhOUmZlVkhOUmZlVkhOUmZlVkhOUmZlVkhOUmZlVkhOUmZlVUuUSlKT9Jd0taaakk7odj5mZdUelEpSkYcD/AG8DxgJHSBrb3ajMzKwbKpWggN2AmRFxX0Q8C/wEOKjLMZmZWRcM73YAdUYBDxaGZwO7FytIGg+Mz4NPSLp7kGKz1kYAD3c7iDXdp7odgFXVy8uq5+MAAAeXSURBVLsdQCdULUGpQVksMxAxCZg0OOFYWZKmRkRvt+MwG4okTe12DJ1QtSa+2cC2heFtgDldisXMzLqoagnqT8COkl4iaR3gcOCyLsdkZmZdUKkmvohYLOkE4CpgGPD9iJjR5bCsHDe7mnXPGrn/KSLa1zIzMxtkVWviMzMzA5ygzMysopygKkTSEknTJd0h6ReSNm1Tf0tJN0m6RdIbByvOFSXp65LetIrzWE/SHyXdKmmGpNMK474i6c+SbpN0SXG9SfonSTfkaW6XtF4uv0bSZqsSk61ZJH0ubye35f1w91z+fwP1RBtJsySNaFPn5JWY7zGSzm4y7mBJp67oPBvM5xxJ9+d1M13SuFwuSd/Mj6e7TdIuhWkaPrpO0lmS3tJ2oRHhv4r8AU8UXk8GPtem/uHA5BVcxrBBfk+bAzcOwHwEbJRfrw3cBOyRh98KDM+vvwR8Kb8eDtwGvDoPb9H//oGj261f/w2dP+B1wA3Aunl4BLB1B5YzCxjRps4TKzHfY4Czm4y7vt0ySy7jHOCQBuUHAFfkfXQP4KZcPgy4F9gBWAe4FRibx20PXN1umb6Cqq4bSE/WQNJLJV0paZqk30l6RT57+TJwQD6bWV/SW/PVws2Sfippozz9LEmnSvo9cGibeqfl8tslvSKXbyTpB7nsNknvzuUN51PnEODK/oFmy2gnkify4Nr5L/K4qyNicR53I+n7c5AS120RcWuu90hELMnjLgOOKLNsGxJGAg9HxCKAiHg4IuYASKpJ6s2vn5D0pbwvXiNptzz+PkkH5jrLXM1IulxSX/0CJV2a5zND6Qk5SDoTWD/v0+flsvfn1oPpkr6j9MxSJB0r6R5JvwH2bPSmJI0BFkXEw3n4nHy1c32O+ZABWHcHAefmffRGYFNJI2nx6LqIeADYQtJWrWbsBFVBeQPcmxe+AzYJ+GhE7Ar8G/CtiJgOnApcEBHjgA2BU4B9ImIXYCrLPhnnmYh4A3BNm3oP5/L/zcsC+A/gsYh4VUT8E/Dr3EzRaj799gSm1ZUttwxJexWaDop/1xfXi6TpwHxgSkTc1GB5x5HO5gDGACHpqpwQP91fKSIWAOtK2qLBPGzouRrYNh/wvyXpzU3qbQjU8r64EDgd2Bd4J/CFFVzmcXk+vcDHJG0REScBT0fEuIh4n6RXAu8B9sz7+RLgfTkBnEbav/YlPVy7kT2Bm+vKRgJvAN4OnAkgaeMm+990Ldu8eUY+SZ0oad1c1ugRdaNalPe7mSaJtV+lvgdl6cwJGE06qE/JVyWvB34qLX0S1LoNpt2DtJH+Iddbh3QV1u+CkvUuzv+nAe/Kr/chNScC6eAu6e1t5tNvJPD3urLllhER1wHjGky/VL76Gad0j+kSSTtHxB394yV9DlgMnJeLhpN2xNcCTwHXSpoWEdfm8fOBrYFHWi3X1nwR8YSkXYE3AnsBF0g6KSLOqav6LC+0CNxOujp5TtLtpP12RXxM0jvz622BHVl+W9wb2BX4U97P1idtt7uTEuXfASRdQDohq9do/7s0Ip4H7pTUAxARC2mz/wGfBR4i7euTgM+QknKzR9S1e3Rd//7XlBNUtTwdEeMkvQi4HPgIqd330Xz21IpIVxXNmq2eLFlvUf6/hBe2D1H3TMQS8+n3NLBeu2VI2guY2GD6pyLi9cWCiHhUUg3YH7gjT3806Yxw78iN3KQztt8Umjd+BewC9Ceo9XJ8Zv0nQDWglhPO0aT9r+i5wvb1PHlbjojnJfXvL4tZtnWqfvsnN/ntA7wuIp7K2/Ny9Uj72eSI+Gzd9Aez/D7ZyNPAi+rKFhVeK89vY+B3Tebx3oi4MyLm9k8v6Qe80MLS7BF16zQp79d2/3MTXwVFxGPAx0gbwNPA/ZIOhaU9Zl7dYLIbgT0lvSzX2yC3P69svaKrgRP6B5R6v5Wdz13Ay9rMn4i4Ljdr1P+9Ps9/y3zlhKT1STv3n/Pw/qSzuQMj4qnCbK8C/inHNhx4M3BnnkbAVqSb1jbESXq5pB0LReOAB1ZydrNIV/prSdqWdC+m3ouABTk5vYLUstHvOUlr59fXAodIenGOc3NJ25M6CfVJ2iLXPbRJLGX3v4VN9r9xEdG/z4zM/wUcTD45JN2KOCofm/Yg3Q6YS/tH140pzKMhJ6iKiohbSL1eDgfeBxwv6VZgBg1+Iytf6h8DnC/pNlICWa4DQtl6dU4HNlPq/n4rsNcKzOeXQF+b+ZcxErguL+tPpKu3y/O4s4GNSU2i0yV9G5beZ/parj8duDkifpmn2ZXUu3AxZrARMFnSnXkbGwtMWMl5/QG4n9QEeBbL3wOC1Ew4PC/ri6T9p98k4DZJ5+XkcApwda47BRiZE8AEUrP6NU2WAfBb4DUq3B9YBeflK8vbSb0cT8/lvwLuA2YC3wX+FdKj60gntleREuWFkR9dl5Pqy0j3rpvyo46s45R6D749Ih7tdiz9JH0DuKxwP8psjZS39V9ExDXdjqVfvve2S0T8R6t6voKywXAisF23g6hzh5OTDRH/CWzQ7SDqDAe+2q6Sr6DMzKySfAVlZmaV5ARlZmaV5ARlZmaV5ARlZmaV5ARlZmaV9P9hQFkOuemocQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig, ax = plt.subplots()\n",
    "ax.bar(np.arange(2), means, align='center', alpha=0.5, ecolor='black', capsize=10)\n",
    "ax.set_ylabel('Mean Mutation Burden')\n",
    "ax.set_xticks(np.arange(2))\n",
    "ax.set_xticklabels([\"Reference \" + \"(n=\" + str(len(ref_mut_burden)) +\")\", \"Simulated \" + \"(n=\" + str(len(simulated_mut_burden)) +\")\"]) \n",
    "ax.set_title('Comparison of Reference and Simulated Mutation Burden')\n",
    "\n",
    "ax.yaxis.grid(True)\n",
    "\n",
    "\n",
    "\n",
    "# Save the figure and show\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Unequal variances T-test\n",
    "\n",
    "\n",
    "Welch's t-test, or unequal variances t-test, is a two-sample location test which is used to test the hypothesis that two populations have equal means.  More reliable when the two samples have unequal variances and/or unequal sample sizes. \n",
    "\n",
    "Accept the null hypothesis that there is no difference between means."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(12475.389821724011, nan)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#General rule: If variance ratio between two populations is less than 4:1, conduct two sample T-test, \n",
    "#else conduct Welch's T Test\n",
    "\n",
    "(np.var(ref_mut_burden), np.var(simulated_mut_burden))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Ttest_indResult(statistic=nan, pvalue=nan)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stats.ttest_ind(simulated_mut_burden, ref_mut_burden, equal_var=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Magnitude Spectra Significance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "kmer_ref = (glob.glob(\"/Users/davidchen/Documents/GitHub/SomaticSiMu/kmer_ref_count/3-mer/3-mer_chr*\"))\n",
    "kmer_count = pd.read_csv(kmer_ref[0], index_col=0)['count'].fillna(0)\n",
    "for i in kmer_ref[1:-1]:\n",
    "    sample = pd.read_csv(i, index_col=0)['count'].fillna(0)\n",
    "    kmer_count = kmer_count + sample\n",
    "\n",
    "kmer_reference_count_dict = dict(zip(pd.read_csv(kmer_ref[0], index_col=0)[\"3\"], kmer_count))\n",
    "ref_genome_length = sum(kmer_reference_count_dict.values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'CTA': 37835932,\n",
       " 'TAA': 45597483,\n",
       " 'AAC': 47201257,\n",
       " 'ACC': 44545092,\n",
       " 'CCC': 48300023,\n",
       " 'CCT': 45730577,\n",
       " 'CTC': 47302637,\n",
       " 'TCT': 50948756,\n",
       " 'CTG': 52003277,\n",
       " 'TGA': 43214327,\n",
       " 'GAA': 46533925,\n",
       " 'AAA': 60869629,\n",
       " 'AAG': 48984923,\n",
       " 'AGT': 44588381,\n",
       " 'GTG': 43579729,\n",
       " 'TGG': 50679722,\n",
       " 'GGA': 43968250,\n",
       " 'GAC': 49542535,\n",
       " 'TAT': 52679409,\n",
       " 'ATC': 45164540,\n",
       " 'TCA': 41218974,\n",
       " 'CAG': 40991591,\n",
       " 'AGC': 47212118,\n",
       " 'GCA': 42425256,\n",
       " 'AGG': 44019928,\n",
       " 'GAT': 45151568,\n",
       " 'ATG': 55474454,\n",
       " 'TGT': 47088446,\n",
       " 'GGG': 40578029,\n",
       " 'GGT': 53518593,\n",
       " 'GAG': 48853223,\n",
       " 'AGA': 44175222,\n",
       " 'ATT': 50532168,\n",
       " 'TTA': 43695544,\n",
       " 'TAG': 44808019,\n",
       " 'AAT': 54457890,\n",
       " 'ATA': 43606596,\n",
       " 'ACT': 42819269,\n",
       " 'TGC': 41195213,\n",
       " 'GCC': 42141983,\n",
       " 'CCA': 45063698,\n",
       " 'GGC': 48080493,\n",
       " 'CAA': 39584962,\n",
       " 'GTC': 43506794,\n",
       " 'TCC': 37295203,\n",
       " 'CTT': 44115524,\n",
       " 'TTT': 44717254,\n",
       " 'TTC': 44741472,\n",
       " 'CAT': 38376928,\n",
       " 'TAC': 37357458,\n",
       " 'GCT': 42321815,\n",
       " 'TCG': 32072699,\n",
       " 'CGT': 24851341,\n",
       " 'GTT': 32507855,\n",
       " 'CAC': 38839731,\n",
       " 'TTG': 36705910,\n",
       " 'ACA': 31256161,\n",
       " 'CCG': 44868097,\n",
       " 'CGC': 25374310,\n",
       " 'CGG': 26182075,\n",
       " 'ACG': 30689434,\n",
       " 'GCG': 34995429,\n",
       " 'GTA': 31265641,\n",
       " 'CGA': 38383658}"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kmer_reference_count_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_ref_spectra(cancer_type, num_data, std_outlier, input_seq_length):\n",
    "    #outliers = outlier_detection(cancer_type, num_data, std_outlier)\n",
    "    bar = num_data[num_data['Cancer Types'] == cancer_type]\n",
    "    #bar = bar[~bar['Sample Names'].isin(outliers)]\n",
    "    ref_mut_burden = list(bar.iloc[:, 3:].sum(axis=1))\n",
    "    ref_mut_burden = [i * (input_seq_length/ref_genome_length) for i in ref_mut_burden]\n",
    "    return ref_mut_burden"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mut_type(cancer_type, simulation_type, gen_start, gen_end, mtype, seq_len):\n",
    "    \n",
    "    if mtype == \"sbs\":\n",
    "        mutation_spectra = pd.DataFrame(index = range(96))\n",
    "        mutation_spectra['Type'] = sbs_prop_data['Type']\n",
    "        mutation_spectra['SubType'] = sbs_prop_data['SubType']\n",
    "\n",
    "    for gen in range(gen_start, gen_end+1):\n",
    "        try:\n",
    "            sbs_spectra = pd.read_csv(abs_path(cancer_type + '_' + str(simulation_type) + '_Stage_Lineage_' + str(gen) + '_' + \"sbs\" + '_freq_table.csv', 'File'))['Frequency']        \n",
    "            sbs_spectra = sbs_spectra.div(seq_len)\n",
    "            mutation_spectra[str(gen)] = sbs_spectra\n",
    "           \n",
    "        except:\n",
    "            print(gen)\n",
    "        \n",
    "    return mutation_spectra"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "1\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "10\n",
      "11\n",
      "12\n",
      "13\n",
      "14\n",
      "15\n",
      "16\n",
      "17\n",
      "18\n",
      "19\n",
      "20\n",
      "21\n",
      "22\n",
      "23\n",
      "24\n",
      "25\n",
      "26\n",
      "27\n",
      "28\n",
      "29\n",
      "30\n",
      "31\n",
      "32\n",
      "33\n",
      "34\n",
      "35\n",
      "36\n",
      "37\n",
      "38\n",
      "39\n",
      "40\n",
      "41\n",
      "42\n",
      "43\n",
      "44\n",
      "45\n",
      "46\n",
      "47\n",
      "48\n",
      "49\n",
      "50\n",
      "51\n",
      "52\n",
      "53\n",
      "54\n",
      "55\n",
      "56\n",
      "57\n",
      "58\n",
      "59\n",
      "60\n",
      "61\n",
      "62\n",
      "63\n",
      "64\n",
      "65\n",
      "66\n",
      "67\n",
      "68\n",
      "69\n",
      "70\n",
      "71\n",
      "72\n",
      "73\n",
      "74\n",
      "75\n",
      "76\n",
      "77\n",
      "78\n",
      "79\n",
      "80\n",
      "81\n",
      "82\n",
      "83\n",
      "84\n",
      "85\n",
      "86\n",
      "87\n",
      "88\n",
      "89\n",
      "90\n",
      "91\n",
      "92\n",
      "93\n",
      "94\n",
      "95\n",
      "96\n",
      "97\n",
      "98\n",
      "99\n",
      "100\n",
      "101\n",
      "102\n",
      "103\n",
      "104\n",
      "105\n",
      "106\n",
      "107\n",
      "108\n",
      "109\n",
      "110\n",
      "111\n",
      "112\n",
      "113\n",
      "114\n",
      "115\n",
      "116\n",
      "117\n",
      "118\n",
      "119\n",
      "120\n",
      "121\n",
      "122\n",
      "123\n",
      "124\n",
      "125\n",
      "126\n",
      "127\n",
      "128\n",
      "129\n",
      "130\n",
      "131\n",
      "132\n",
      "133\n",
      "134\n",
      "135\n",
      "136\n",
      "137\n",
      "138\n",
      "139\n",
      "140\n",
      "141\n",
      "142\n",
      "143\n",
      "144\n",
      "145\n",
      "146\n",
      "147\n",
      "148\n",
      "149\n",
      "150\n",
      "151\n",
      "152\n",
      "153\n",
      "154\n",
      "155\n",
      "156\n",
      "157\n",
      "158\n",
      "159\n",
      "160\n",
      "161\n",
      "162\n",
      "163\n",
      "164\n",
      "165\n",
      "166\n",
      "167\n",
      "168\n",
      "169\n",
      "170\n",
      "171\n",
      "172\n",
      "173\n",
      "174\n",
      "175\n",
      "176\n",
      "177\n",
      "178\n",
      "179\n",
      "180\n",
      "181\n",
      "182\n",
      "183\n",
      "184\n",
      "185\n",
      "186\n",
      "187\n",
      "188\n",
      "189\n",
      "190\n",
      "191\n",
      "192\n",
      "193\n",
      "194\n",
      "195\n",
      "196\n",
      "197\n",
      "198\n",
      "199\n",
      "200\n",
      "201\n",
      "202\n",
      "203\n",
      "204\n",
      "205\n",
      "206\n",
      "207\n",
      "208\n",
      "209\n",
      "210\n",
      "211\n",
      "212\n",
      "213\n",
      "214\n",
      "215\n",
      "216\n",
      "217\n",
      "218\n",
      "219\n",
      "220\n",
      "221\n",
      "222\n",
      "223\n",
      "224\n",
      "225\n",
      "226\n",
      "227\n",
      "228\n",
      "229\n",
      "230\n",
      "231\n",
      "232\n",
      "233\n",
      "234\n",
      "235\n",
      "236\n",
      "237\n",
      "238\n",
      "239\n",
      "240\n",
      "241\n",
      "242\n",
      "243\n",
      "244\n",
      "245\n",
      "246\n",
      "247\n",
      "248\n",
      "249\n",
      "250\n",
      "251\n",
      "252\n",
      "253\n",
      "254\n",
      "255\n",
      "256\n",
      "257\n",
      "258\n",
      "259\n",
      "260\n",
      "261\n",
      "262\n",
      "263\n",
      "264\n",
      "265\n",
      "266\n",
      "267\n",
      "268\n",
      "269\n",
      "270\n",
      "271\n",
      "272\n",
      "273\n",
      "274\n",
      "275\n",
      "276\n",
      "277\n",
      "278\n",
      "279\n",
      "280\n",
      "281\n",
      "282\n",
      "283\n",
      "284\n",
      "285\n",
      "286\n",
      "287\n",
      "288\n",
      "289\n",
      "290\n",
      "291\n",
      "292\n",
      "293\n",
      "294\n",
      "295\n",
      "296\n",
      "297\n",
      "298\n",
      "299\n",
      "300\n",
      "301\n",
      "302\n",
      "303\n",
      "304\n",
      "305\n",
      "306\n",
      "307\n",
      "308\n",
      "309\n",
      "310\n",
      "311\n",
      "312\n",
      "313\n",
      "314\n",
      "315\n",
      "316\n",
      "317\n",
      "318\n",
      "319\n",
      "320\n",
      "321\n",
      "322\n",
      "323\n",
      "324\n",
      "325\n",
      "326\n",
      "327\n",
      "328\n",
      "329\n",
      "330\n",
      "331\n",
      "332\n",
      "333\n",
      "334\n",
      "335\n",
      "336\n",
      "337\n",
      "338\n",
      "339\n",
      "340\n",
      "341\n",
      "342\n",
      "343\n",
      "344\n",
      "345\n",
      "346\n",
      "347\n",
      "348\n",
      "349\n",
      "350\n",
      "351\n",
      "352\n",
      "353\n",
      "354\n",
      "355\n",
      "356\n",
      "357\n",
      "358\n",
      "359\n",
      "360\n",
      "361\n",
      "362\n",
      "363\n",
      "364\n",
      "365\n",
      "366\n",
      "367\n",
      "368\n",
      "369\n",
      "370\n",
      "371\n",
      "372\n",
      "373\n",
      "374\n",
      "375\n",
      "376\n",
      "377\n",
      "378\n",
      "379\n",
      "380\n",
      "381\n",
      "382\n",
      "383\n",
      "384\n",
      "385\n",
      "386\n",
      "387\n",
      "388\n",
      "389\n",
      "390\n",
      "391\n",
      "392\n",
      "393\n",
      "394\n",
      "395\n",
      "396\n",
      "397\n",
      "398\n",
      "399\n",
      "400\n",
      "401\n",
      "402\n",
      "403\n",
      "404\n",
      "405\n",
      "406\n",
      "407\n",
      "408\n",
      "409\n",
      "410\n",
      "411\n",
      "412\n",
      "413\n",
      "414\n",
      "415\n",
      "416\n",
      "417\n",
      "418\n",
      "419\n",
      "420\n",
      "421\n",
      "422\n",
      "423\n",
      "424\n",
      "425\n",
      "426\n",
      "427\n",
      "428\n",
      "429\n",
      "430\n",
      "431\n",
      "432\n",
      "433\n",
      "434\n",
      "435\n",
      "436\n",
      "437\n",
      "438\n",
      "439\n",
      "440\n",
      "441\n",
      "442\n",
      "443\n",
      "444\n",
      "445\n",
      "446\n",
      "447\n",
      "448\n",
      "449\n",
      "450\n",
      "451\n",
      "452\n",
      "453\n",
      "454\n",
      "455\n",
      "456\n",
      "457\n",
      "458\n",
      "459\n",
      "460\n",
      "461\n",
      "462\n",
      "463\n",
      "464\n",
      "465\n",
      "466\n",
      "467\n",
      "468\n",
      "469\n",
      "470\n",
      "471\n",
      "472\n",
      "473\n",
      "474\n",
      "475\n",
      "476\n",
      "477\n",
      "478\n",
      "479\n",
      "480\n",
      "481\n",
      "482\n",
      "483\n",
      "484\n",
      "485\n",
      "486\n",
      "487\n",
      "488\n",
      "489\n",
      "490\n",
      "491\n",
      "492\n",
      "493\n",
      "494\n",
      "495\n",
      "496\n",
      "497\n",
      "498\n",
      "499\n"
     ]
    }
   ],
   "source": [
    "sim_mut_spectra = mut_type(\"Liver-HCC\", simulation_type=\"End\", gen_start=0, gen_end=499, mtype=\"sbs\", seq_len=50818467)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Type</th>\n",
       "      <th>SubType</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C&gt;A</td>\n",
       "      <td>ACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C&gt;A</td>\n",
       "      <td>ACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C&gt;A</td>\n",
       "      <td>ACG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C&gt;A</td>\n",
       "      <td>ACT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C&gt;A</td>\n",
       "      <td>CCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>T&gt;G</td>\n",
       "      <td>GTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>T&gt;G</td>\n",
       "      <td>TTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>T&gt;G</td>\n",
       "      <td>TTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>T&gt;G</td>\n",
       "      <td>TTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>T&gt;G</td>\n",
       "      <td>TTT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>96 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Type SubType\n",
       "0   C>A     ACA\n",
       "1   C>A     ACC\n",
       "2   C>A     ACG\n",
       "3   C>A     ACT\n",
       "4   C>A     CCA\n",
       "..  ...     ...\n",
       "91  T>G     GTT\n",
       "92  T>G     TTA\n",
       "93  T>G     TTC\n",
       "94  T>G     TTG\n",
       "95  T>G     TTT\n",
       "\n",
       "[96 rows x 2 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sim_mut_spectra"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_mut_spectra = mut_prob(cancer_type=\"Liver-HCC\", prop_data=sbs_prop_data, num_data=sbs_num_data, freq_folder_path=sbs_freq_folder_path, mut_type=\"SBS\", std_outlier=3).iloc[:, 2:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SBS1</th>\n",
       "      <th>SBS2</th>\n",
       "      <th>SBS3</th>\n",
       "      <th>SBS4</th>\n",
       "      <th>SBS5</th>\n",
       "      <th>SBS6</th>\n",
       "      <th>SBS7a</th>\n",
       "      <th>SBS7b</th>\n",
       "      <th>SBS7c</th>\n",
       "      <th>SBS7d</th>\n",
       "      <th>...</th>\n",
       "      <th>SBS51</th>\n",
       "      <th>SBS52</th>\n",
       "      <th>SBS53</th>\n",
       "      <th>SBS54</th>\n",
       "      <th>SBS55</th>\n",
       "      <th>SBS56</th>\n",
       "      <th>SBS57</th>\n",
       "      <th>SBS58</th>\n",
       "      <th>SBS59</th>\n",
       "      <th>SBS60</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2.151708e-09</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.218182e-07</td>\n",
       "      <td>2.610684e-06</td>\n",
       "      <td>6.422036e-10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.104609e-10</td>\n",
       "      <td>1.945248e-10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.082993e-10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4.691692e-09</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.163426e-07</td>\n",
       "      <td>1.740164e-06</td>\n",
       "      <td>6.709038e-10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.409789e-11</td>\n",
       "      <td>6.074065e-11</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.533879e-10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4.629759e-10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.670228e-07</td>\n",
       "      <td>4.334913e-07</td>\n",
       "      <td>8.462968e-11</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.425327e-10</td>\n",
       "      <td>1.590745e-10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.423326e-11</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2.764616e-09</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.730997e-07</td>\n",
       "      <td>1.278939e-06</td>\n",
       "      <td>2.418977e-10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.122466e-13</td>\n",
       "      <td>3.315866e-11</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.312215e-09</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5.916712e-10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.376520e-06</td>\n",
       "      <td>1.262229e-06</td>\n",
       "      <td>2.147492e-09</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.354871e-09</td>\n",
       "      <td>4.444415e-12</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.242309e-10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>3.543170e-11</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.500764e-09</td>\n",
       "      <td>5.108937e-07</td>\n",
       "      <td>1.188360e-09</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.164970e-11</td>\n",
       "      <td>1.088921e-12</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.307960e-12</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>8.054566e-22</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.224791e-08</td>\n",
       "      <td>1.689007e-06</td>\n",
       "      <td>2.359722e-10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.969697e-11</td>\n",
       "      <td>4.339643e-12</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.025694e-10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>1.069796e-10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.038658e-09</td>\n",
       "      <td>1.140977e-06</td>\n",
       "      <td>3.467093e-10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.156695e-11</td>\n",
       "      <td>5.083066e-11</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.128632e-14</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>2.098312e-09</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.510570e-08</td>\n",
       "      <td>2.237611e-06</td>\n",
       "      <td>7.255726e-10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.871343e-10</td>\n",
       "      <td>1.026360e-09</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.602297e-12</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>6.436191e-22</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.474537e-08</td>\n",
       "      <td>3.490450e-06</td>\n",
       "      <td>1.813759e-09</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.720243e-10</td>\n",
       "      <td>5.340677e-11</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.210926e-09</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>96 rows × 65 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            SBS1  SBS2  SBS3          SBS4          SBS5          SBS6  SBS7a  \\\n",
       "0   2.151708e-09   0.0   0.0  9.218182e-07  2.610684e-06  6.422036e-10    0.0   \n",
       "1   4.691692e-09   0.0   0.0  6.163426e-07  1.740164e-06  6.709038e-10    0.0   \n",
       "2   4.629759e-10   0.0   0.0  3.670228e-07  4.334913e-07  8.462968e-11    0.0   \n",
       "3   2.764616e-09   0.0   0.0  5.730997e-07  1.278939e-06  2.418977e-10    0.0   \n",
       "4   5.916712e-10   0.0   0.0  1.376520e-06  1.262229e-06  2.147492e-09    0.0   \n",
       "..           ...   ...   ...           ...           ...           ...    ...   \n",
       "91  3.543170e-11   0.0   0.0  5.500764e-09  5.108937e-07  1.188360e-09    0.0   \n",
       "92  8.054566e-22   0.0   0.0  1.224791e-08  1.689007e-06  2.359722e-10    0.0   \n",
       "93  1.069796e-10   0.0   0.0  3.038658e-09  1.140977e-06  3.467093e-10    0.0   \n",
       "94  2.098312e-09   0.0   0.0  7.510570e-08  2.237611e-06  7.255726e-10    0.0   \n",
       "95  6.436191e-22   0.0   0.0  1.474537e-08  3.490450e-06  1.813759e-09    0.0   \n",
       "\n",
       "    SBS7b  SBS7c  SBS7d  ...  SBS51  SBS52         SBS53         SBS54  SBS55  \\\n",
       "0     0.0    0.0    0.0  ...    0.0    0.0  1.104609e-10  1.945248e-10    0.0   \n",
       "1     0.0    0.0    0.0  ...    0.0    0.0  3.409789e-11  6.074065e-11    0.0   \n",
       "2     0.0    0.0    0.0  ...    0.0    0.0  8.425327e-10  1.590745e-10    0.0   \n",
       "3     0.0    0.0    0.0  ...    0.0    0.0  3.122466e-13  3.315866e-11    0.0   \n",
       "4     0.0    0.0    0.0  ...    0.0    0.0  5.354871e-09  4.444415e-12    0.0   \n",
       "..    ...    ...    ...  ...    ...    ...           ...           ...    ...   \n",
       "91    0.0    0.0    0.0  ...    0.0    0.0  4.164970e-11  1.088921e-12    0.0   \n",
       "92    0.0    0.0    0.0  ...    0.0    0.0  3.969697e-11  4.339643e-12    0.0   \n",
       "93    0.0    0.0    0.0  ...    0.0    0.0  7.156695e-11  5.083066e-11    0.0   \n",
       "94    0.0    0.0    0.0  ...    0.0    0.0  1.871343e-10  1.026360e-09    0.0   \n",
       "95    0.0    0.0    0.0  ...    0.0    0.0  1.720243e-10  5.340677e-11    0.0   \n",
       "\n",
       "           SBS56  SBS57  SBS58  SBS59  SBS60  \n",
       "0   8.082993e-10    0.0    0.0    0.0    0.0  \n",
       "1   8.533879e-10    0.0    0.0    0.0    0.0  \n",
       "2   1.423326e-11    0.0    0.0    0.0    0.0  \n",
       "3   1.312215e-09    0.0    0.0    0.0    0.0  \n",
       "4   1.242309e-10    0.0    0.0    0.0    0.0  \n",
       "..           ...    ...    ...    ...    ...  \n",
       "91  7.307960e-12    0.0    0.0    0.0    0.0  \n",
       "92  9.025694e-10    0.0    0.0    0.0    0.0  \n",
       "93  5.128632e-14    0.0    0.0    0.0    0.0  \n",
       "94  9.602297e-12    0.0    0.0    0.0    0.0  \n",
       "95  2.210926e-09    0.0    0.0    0.0    0.0  \n",
       "\n",
       "[96 rows x 65 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_mut_spectra"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "cosine_sim_96 = pd.DataFrame(index=range(500), columns=list(ref_mut_spectra.columns))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "ename": "IndexError",
     "evalue": "single positional indexer is out-of-bounds",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-31-e5568d0f65c6>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m500\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mj\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mcosine_sim_96\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m         \u001b[0mcosine_sim_96\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mj\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcosine_similarity\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mref_mut_spectra\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mj\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msim_mut_spectra\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m2\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/opt/anaconda3/lib/python3.8/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   1760\u001b[0m                 \u001b[0;32mexcept\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mKeyError\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mIndexError\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mAttributeError\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1761\u001b[0m                     \u001b[0;32mpass\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1762\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_tuple\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1763\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1764\u001b[0m             \u001b[0;31m# we by definition only have the 0th axis\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/lib/python3.8/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_getitem_tuple\u001b[0;34m(self, tup)\u001b[0m\n\u001b[1;32m   2065\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_getitem_tuple\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtup\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mTuple\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2066\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2067\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_has_valid_tuple\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtup\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2068\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2069\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_lowerdim\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtup\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/lib/python3.8/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_has_valid_tuple\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    701\u001b[0m                 \u001b[0;32mraise\u001b[0m \u001b[0mIndexingError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Too many indexers\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    702\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 703\u001b[0;31m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_validate_key\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    704\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    705\u001b[0m                 raise ValueError(\n",
      "\u001b[0;32m/opt/anaconda3/lib/python3.8/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_validate_key\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1992\u001b[0m             \u001b[0;32mreturn\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1993\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0mis_integer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1994\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_validate_integer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1995\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtuple\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1996\u001b[0m             \u001b[0;31m# a tuple should already have been caught by this point\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/lib/python3.8/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_validate_integer\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   2061\u001b[0m         \u001b[0mlen_axis\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2062\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mkey\u001b[0m \u001b[0;34m>=\u001b[0m \u001b[0mlen_axis\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mkey\u001b[0m \u001b[0;34m<\u001b[0m \u001b[0;34m-\u001b[0m\u001b[0mlen_axis\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2063\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mIndexError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"single positional indexer is out-of-bounds\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2064\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2065\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_getitem_tuple\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtup\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mTuple\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mIndexError\u001b[0m: single positional indexer is out-of-bounds"
     ]
    }
   ],
   "source": [
    "for i in range(500):\n",
    "    for j in cosine_sim_96.columns:\n",
    "        cosine_sim_96.loc[i,j] = np.min(cosine_similarity([list(ref_mut_spectra.loc[:, j]), list(sim_mut_spectra.iloc[:, 2:].iloc[:,i])]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "cosine_sim_96"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cosine_sim_96.mean(axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "cosine_sim_96.mean(axis=0).nlargest(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "network_df = pd.DataFrame(index=range(len(cosine_sim_96.columns)), columns = [\"node1\", \"node2\", \"value\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "network_df['node1'] = \"Liver-HCC\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "network_df['node2'] = cosine_sim_96.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "network_df['value'] = list(cosine_sim_96.mean(axis=0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "network_df[network_df['value']>0.3].sort_values('value')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Keep only correlation over a threshold and remove self correlation (cor(A,A)=1)\n",
    "links_filtered=network_df.loc[ (network_df['value'] > 0.3) & (network_df['node1'] != network_df['node2']) ]\n",
    "links_filtered\n",
    " \n",
    "plt.figure(2,figsize=(15,12)) \n",
    "# Build your graph\n",
    "G=nx.from_pandas_edgelist(links_filtered, 'node1', 'node2')\n",
    "\n",
    "# Plot the network:\n",
    "#colors = [i/len(G.nodes) for i in range(len(G.nodes))]\n",
    "colors = network_df[network_df['value']>0.3]['value'].tolist()\n",
    "colors.insert(0,0.1)\n",
    "nx.draw_spring(G, with_labels=True, node_color=colors ,cmap=plt.cm.YlOrRd, node_size=800,linewidths=50, font_size=20, width=5)\n",
    "\n",
    "vmin = network_df['value'].min()\n",
    "vmax = network_df['value'].max()\n",
    "\n",
    "sm = plt.cm.ScalarMappable(cmap=\"YlOrRd\", norm=plt.Normalize(vmin=network_df['value'].min(), vmax=network_df['value'].max()))\n",
    "sm.set_array([])\n",
    "cbar = plt.colorbar(sm, anchor=(0.2,0.0))\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sim_mut_spectra2 = mut_type(\"Biliary-AdenoCA\", simulation_type=\"End\", gen_start=0, gen_end=499, mtype=\"sbs\", seq_len=50818467)\n",
    "ref_mut_spectra2 = mut_prob(cancer_type=\"Biliary-AdenoCA\", prop_data=sbs_prop_data, num_data=sbs_num_data, freq_folder_path=sbs_freq_folder_path, mut_type=\"SBS\", std_outlier=3).iloc[:, 2:]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cosine_sim_962 = pd.DataFrame(index=range(500), columns=list(ref_mut_spectra2.columns))\n",
    "for i in range(500):\n",
    "    for j in cosine_sim_962.columns:\n",
    "        cosine_sim_962.loc[i,j] = np.min(cosine_similarity([list(ref_mut_spectra2.loc[:, j]), list(sim_mut_spectra2.iloc[:, 2:].iloc[:,i])]))\n",
    "        \n",
    "network_df2 = pd.DataFrame(index=range(len(cosine_sim_962.columns)), columns = [\"node1\", \"node2\", \"value\"])\n",
    "\n",
    "network_df2['node1'] = \"Biliary-AdenoCA\"\n",
    "network_df2['node2'] = cosine_sim_962.columns\n",
    "network_df2['value'] = list(cosine_sim_962.mean(axis=0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "combined_df = network_df.append(network_df2)\n",
    "combined_df.reset_index(inplace=True, drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Keep only correlation over a threshold and remove self correlation (cor(A,A)=1)\n",
    "links_filtered2=combined_df.loc[ (combined_df['value'] > 0.3) & (combined_df['node1'] != combined_df['node2']) ]\n",
    "links_filtered2\n",
    " \n",
    "plt.figure(2,figsize=(15,12)) \n",
    "# Build your graph\n",
    "G=nx.from_pandas_edgelist(links_filtered2, 'node1', 'node2')\n",
    "degrees = G.degree()\n",
    "# Plot the network:\n",
    "#colors = [i/15 for i in range(15)]\n",
    "edge_colors = combined_df[combined_df['value']>0.3]['value'].tolist()\n",
    "node_colors = [\"silver\" if n in [\"Liver-HCC\", \"Biliary-AdenoCA\"] else \"whitesmoke\" for n in G.nodes()]\n",
    "node_sizes = [15000 if n in [\"Liver-HCC\", \"Biliary-AdenoCA\"] else 1000 for n in G.nodes()]\n",
    "\n",
    "nx.draw_spring(G, with_labels=True, node_color = node_colors, edge_color=edge_colors, edge_cmap=plt.cm.YlOrRd, node_size=node_sizes,linewidths=50, font_size=20, width=8)\n",
    "\n",
    "#node_color\n",
    "\n",
    "#vmin = links_filtered2['value'].min()-(0.1*links_filtered2['value'].mean()) \n",
    "#vmin = vmin - (vmin%0.1)\n",
    "#vmax = links_filtered2['value'].max()+(0.1*links_filtered2['value'].mean()) \n",
    "#vmax = vmax + (0.1 - vmax%0.1)\n",
    "\n",
    "sm = plt.cm.ScalarMappable(cmap=\"YlOrRd\", norm=plt.Normalize(vmin=links_filtered2['value'].min(), vmax=links_filtered2['value'].max()))\n",
    "sm.set_array([])\n",
    "cbar = plt.colorbar(sm, anchor=(0.2,0.0))\n",
    "\n",
    "plt.title(\"Cosine Similarity Network\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
